// ng-primitives/esm2022/pagination/index.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
export { NgpPagination } from './pagination/pagination.directive';
export { NgpPaginationToken } from './pagination/pagination.token';
export { NgpPaginationButton } from './pagination-button/pagination-button.directive';
export { NgpPaginationButtonToken } from './pagination-button/pagination-button.token';
export { NgpPaginationPrevious } from './pagination-previous/pagination-previous.directive';
export { NgpPaginationPreviousToken } from './pagination-previous/pagination-previous.token';
export { NgpPaginationNext } from './pagination-next/pagination-next.directive';
export { NgpPaginationNextToken } from './pagination-next/pagination-next.token';
export { NgpPaginationFirst } from './pagination-first/pagination-first.directive';
export { NgpPaginationFirstToken } from './pagination-first/pagination-first.token';
export { NgpPaginationLast } from './pagination-last/pagination-last.directive';
export { NgpPaginationLastToken } from './pagination-last/pagination-last.token';
