// @ng-matero/extensions/esm2022/colorpicker/colorpicker-module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { ColorChromeModule } from 'ngx-color/chrome';
import { MTX_COLORPICKER_SCROLL_STRATEGY_FACTORY_PROVIDER, MtxColorpicker, MtxColorpickerContent, } from './colorpicker';
import { MtxColorpickerInput } from './colorpicker-input';
import { MtxColorpickerToggle, MtxColorpickerToggleIcon } from './colorpicker-toggle';
import * as i0 from "@angular/core";
export class MtxColorpickerModule {
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxColorpickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.0", ngImport: i0, type: MtxColorpickerModule, imports: [CommonModule,
            OverlayModule,
            A11yModule,
            PortalModule,
            MatButtonModule,
            ColorChromeModule,
            MtxColorpicker,
            MtxColorpickerContent,
            MtxColorpickerInput,
            MtxColorpickerToggle,
            MtxColorpickerToggleIcon], exports: [MtxColorpicker,
            MtxColorpickerContent,
            MtxColorpickerInput,
            MtxColorpickerToggle,
            MtxColorpickerToggleIcon] }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxColorpickerModule, providers: [MTX_COLORPICKER_SCROLL_STRATEGY_FACTORY_PROVIDER], imports: [CommonModule,
            OverlayModule,
            A11yModule,
            PortalModule,
            MatButtonModule,
            ColorChromeModule,
            MtxColorpickerContent,
            MtxColorpickerToggle] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxColorpickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        OverlayModule,
                        A11yModule,
                        PortalModule,
                        MatButtonModule,
                        ColorChromeModule,
                        MtxColorpicker,
                        MtxColorpickerContent,
                        MtxColorpickerInput,
                        MtxColorpickerToggle,
                        MtxColorpickerToggleIcon,
                    ],
                    exports: [
                        MtxColorpicker,
                        MtxColorpickerContent,
                        MtxColorpickerInput,
                        MtxColorpickerToggle,
                        MtxColorpickerToggleIcon,
                    ],
                    providers: [MTX_COLORPICKER_SCROLL_STRATEGY_FACTORY_PROVIDER],
                }]
        }] });
