// ng-primitives/esm2022/avatar/avatar-image/avatar-image.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, ElementRef, HostListener, inject } from '@angular/core';
import { NgpVisuallyHidden, injectVisuallyHidden } from 'ng-primitives/a11y';
import { NgpAvatarState } from '../avatar/avatar.directive';
import { injectAvatar } from '../avatar/avatar.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/a11y";
export class NgpAvatarImage {
    constructor() {
        /**
         * Control the visibility of the image.
         */
        this.visuallyHidden = injectVisuallyHidden();
        /**
         * Access the avatar
         */
        this.avatar = injectAvatar();
        /**
         * Access the image element ref.
         */
        this.elementRef = inject(ElementRef);
    }
    ngOnInit() {
        // mark the avatar as loading
        this.setState(NgpAvatarState.Loading);
        // if there is no src, we can report this as an error
        if (!this.elementRef.nativeElement.src) {
            this.setState(NgpAvatarState.Error);
        }
        // if the image has already loaded, we can report this to the avatar
        if (this.elementRef.nativeElement.complete) {
            this.setState(NgpAvatarState.Loaded);
        }
    }
    onLoad() {
        this.setState(NgpAvatarState.Loaded);
    }
    onError() {
        this.setState(NgpAvatarState.Error);
    }
    setState(state) {
        this.avatar.setState(state);
        // if the state is loaded then we should make the image visible
        this.visuallyHidden.setVisibility(state === NgpAvatarState.Loaded);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAvatarImage, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpAvatarImage, isStandalone: true, selector: "img[ngpAvatarImage]", host: { listeners: { "load": "onLoad()", "error": "onError()" } }, hostDirectives: [{ directive: i1.NgpVisuallyHidden }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAvatarImage, decorators: [{
            type: Directive,
            args: [{
                    selector: 'img[ngpAvatarImage]',
                    standalone: true,
                    hostDirectives: [NgpVisuallyHidden],
                }]
        }], propDecorators: { onLoad: [{
                type: HostListener,
                args: ['load']
            }], onError: [{
                type: HostListener,
                args: ['error']
            }] } });
