// @ng-matero/extensions/esm2022/popover/popover-target.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
export class MtxPopoverTarget {
    constructor(elementRef) {
        this.elementRef = elementRef;
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxPopoverTarget, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    /** @nocollapse */ static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.0", type: MtxPopoverTarget, isStandalone: true, selector: "mtx-popover-target, [mtxPopoverTarget]", exportAs: ["mtxPopoverTarget"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxPopoverTarget, decorators: [{
            type: Directive,
            args: [{
                    selector: 'mtx-popover-target, [mtxPopoverTarget]',
                    exportAs: 'mtxPopoverTarget',
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: i0.ElementRef }] });
