// ng-primitives/esm2022/date-picker/date-picker-cell/date-picker-cell.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { computed, contentChild, Directive } from '@angular/core';
import { NgpDatePickerDateButtonToken } from '../date-picker-date-button/date-picker-date-button.token';
import { injectDatePicker } from '../date-picker/date-picker.token';
import { NgpDatePickerCellToken } from './date-picker-cell.token';
import * as i0 from "@angular/core";
export class NgpDatePickerCell {
    constructor() {
        /**
         * Access the date picker.
         */
        this.datePicker = injectDatePicker();
        /**
         * Access the child date picker date button.
         */
        this.datePickerButton = contentChild(NgpDatePickerDateButtonToken, {
            descendants: true,
        });
        /**
         * Access the label id.
         */
        this.labelId = computed(() => this.datePicker.label()?.id());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDatePickerCell, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.2.0", version: "18.2.13", type: NgpDatePickerCell, isStandalone: true, selector: "[ngpDatePickerCell]", host: { attributes: { "role": "gridcell" }, properties: { "attr.data-selected": "datePickerButton()?.selected() ? \"\" : null", "attr.aria-selected": "datePickerButton()?.selected()", "attr.aria-disabled": "datePickerButton()?.disabled()", "attr.data-disabled": "datePickerButton()?.disabled() ? \"\" : null", "attr.aria-labelledby": "labelId()" } }, providers: [{ provide: NgpDatePickerCellToken, useExisting: NgpDatePickerCell }], queries: [{ propertyName: "datePickerButton", first: true, predicate: NgpDatePickerDateButtonToken, descendants: true, isSignal: true }], exportAs: ["ngpDatePickerCell"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDatePickerCell, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpDatePickerCell]',
                    exportAs: 'ngpDatePickerCell',
                    providers: [{ provide: NgpDatePickerCellToken, useExisting: NgpDatePickerCell }],
                    host: {
                        role: 'gridcell',
                        '[attr.data-selected]': 'datePickerButton()?.selected() ? "" : null',
                        '[attr.aria-selected]': 'datePickerButton()?.selected()',
                        '[attr.aria-disabled]': 'datePickerButton()?.disabled()',
                        '[attr.data-disabled]': 'datePickerButton()?.disabled() ? "" : null',
                        '[attr.aria-labelledby]': 'labelId()',
                    },
                }]
        }] });
