// ng-primitives/esm2022/slider/slider-track/slider-track.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, ElementRef, HostListener, inject } from '@angular/core';
import { injectSlider } from '../slider/slider.token';
import { NgpSliderTrackToken } from './slider-track.token';
import * as i0 from "@angular/core";
export class NgpSliderTrack {
    constructor() {
        /**
         * Access the slider.
         */
        this.slider = injectSlider();
        /**
         * The element that represents the slider track.
         */
        this.element = inject(ElementRef);
    }
    /**
     * When the slider track is clicked, update the value.
     * @param event The click event.
     */
    handlePointerDown(event) {
        if (this.slider.disabled()) {
            return;
        }
        // get the position the click occurred within the slider track
        const position = this.slider.orientation() === 'horizontal' ? event.clientX : event.clientY;
        const rect = this.element.nativeElement.getBoundingClientRect();
        const percentage = (position - (this.slider.orientation() === 'horizontal' ? rect.left : rect.top)) /
            (this.slider.orientation() === 'horizontal' ? rect.width : rect.height);
        // update the value based on the position
        this.slider.value.set(this.slider.min() + (this.slider.max() - this.slider.min()) * percentage);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSliderTrack, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpSliderTrack, isStandalone: true, selector: "[ngpSliderTrack]", host: { listeners: { "pointerdown": "handlePointerDown($event)" }, properties: { "attr.data-orientation": "slider.orientation()", "attr.data-disabled": "slider.disabled() ? \"\" : null" } }, providers: [{ provide: NgpSliderTrackToken, useExisting: NgpSliderTrack }], exportAs: ["ngpSliderTrack"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSliderTrack, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpSliderTrack]',
                    exportAs: 'ngpSliderTrack',
                    providers: [{ provide: NgpSliderTrackToken, useExisting: NgpSliderTrack }],
                    host: {
                        '[attr.data-orientation]': 'slider.orientation()',
                        '[attr.data-disabled]': 'slider.disabled() ? "" : null',
                    },
                }]
        }], propDecorators: { handlePointerDown: [{
                type: HostListener,
                args: ['pointerdown', ['$event']]
            }] } });
