// ng-primitives/esm2022/menu/menu-trigger/menu-trigger.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { CdkMenuTrigger } from '@angular/cdk/menu';
import { Directive, inject, signal } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { NgpMenuTriggerToken } from './menu-trigger.token';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/menu";
export class NgpMenuTrigger {
    constructor() {
        /**
         * Access to the underlying `CdkMenuTrigger`.
         */
        this.cdkMenuTrigger = inject(CdkMenuTrigger);
        /**
         * Store the open state of the menu.
         */
        this.open = signal(false);
        this.cdkMenuTrigger.opened.pipe(takeUntilDestroyed()).subscribe(() => this.open.set(true));
        this.cdkMenuTrigger.closed.pipe(takeUntilDestroyed()).subscribe(() => this.open.set(false));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpMenuTrigger, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpMenuTrigger, isStandalone: true, selector: "[ngpMenuTrigger]", host: { properties: { "attr.data-open": "open() ? \"\" : null" } }, providers: [{ provide: NgpMenuTriggerToken, useExisting: NgpMenuTrigger }], exportAs: ["ngpMenuTrigger"], hostDirectives: [{ directive: i1.CdkMenuTrigger, inputs: ["cdkMenuTriggerFor", "ngpMenuTrigger"] }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpMenuTrigger, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpMenuTrigger]',
                    exportAs: 'ngpMenuTrigger',
                    providers: [{ provide: NgpMenuTriggerToken, useExisting: NgpMenuTrigger }],
                    hostDirectives: [{ directive: CdkMenuTrigger, inputs: ['cdkMenuTriggerFor: ngpMenuTrigger'] }],
                    host: {
                        '[attr.data-open]': 'open() ? "" : null',
                    },
                }]
        }], ctorParameters: () => [] });
