// ng-primitives/esm2022/menu/menu/menu.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { CdkMenu } from '@angular/cdk/menu';
import { Directive } from '@angular/core';
import { NgpMenuToken } from './menu.token';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/menu";
export class NgpMenu {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpMenu, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpMenu, isStandalone: true, selector: "[ngpMenu]", providers: [{ provide: NgpMenuToken, useExisting: NgpMenu }], exportAs: ["ngpMenu"], hostDirectives: [{ directive: i1.CdkMenu }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpMenu, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpMenu]',
                    exportAs: 'ngpMenu',
                    providers: [{ provide: NgpMenuToken, useExisting: NgpMenu }],
                    hostDirectives: [CdkMenu],
                }]
        }] });
