// ng-primitives/esm2022/tooltip/tooltip/tooltip.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { isPlatformBrowser } from '@angular/common';
import { Directive, ElementRef, PLATFORM_ID, computed, inject, isDevMode, } from '@angular/core';
import { injectTooltipTrigger } from '../tooltip-trigger/tooltip-trigger.token';
import { NgpTooltipToken } from './tooltip.token';
import * as i0 from "@angular/core";
export class NgpTooltip {
    constructor() {
        /**
         * Access the tooltip element.
         */
        this.tooltip = inject((ElementRef));
        /**
         * Access the platform.
         */
        this.platform = inject(PLATFORM_ID);
        /**
         * Access the trigger instance.
         */
        this.trigger = injectTooltipTrigger();
        /**
         * Compute the x position of the tooltip.
         */
        this.x = computed(() => this.trigger.position().x);
        /**
         * Compute the y position of the tooltip.
         */
        this.y = computed(() => this.trigger.position().y);
    }
    ngOnInit() {
        // if the element does not have a fixed position then throw a warning in dev mode
        if (isDevMode() && isPlatformBrowser(this.platform)) {
            const { position } = getComputedStyle(this.tooltip.nativeElement);
            if (position !== 'absolute') {
                console.warn(`The tooltip element must have an absolute position. The current position is ${position}.`);
            }
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTooltip, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpTooltip, isStandalone: true, selector: "[ngpTooltip]", host: { attributes: { "role": "tooltip" }, properties: { "style.left.px": "x()", "style.top.px": "y()" } }, providers: [{ provide: NgpTooltipToken, useExisting: NgpTooltip }], exportAs: ["ngpTooltip"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTooltip, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpTooltip]',
                    exportAs: 'ngpTooltip',
                    providers: [{ provide: NgpTooltipToken, useExisting: NgpTooltip }],
                    host: {
                        role: 'tooltip',
                        '[style.left.px]': 'x()',
                        '[style.top.px]': 'y()',
                    },
                }]
        }] });
