// ng-primitives/esm2022/dialog/dialog-overlay/dialog-overlay.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, HostListener } from '@angular/core';
import { injectDialogRef } from '../dialog/dialog-ref';
import { NgpDialogOverlayToken } from './dialog-overlay.token';
import * as i0 from "@angular/core";
export class NgpDialogOverlay {
    constructor() {
        /** Access the dialog ref. */
        this.dialogRef = injectDialogRef();
    }
    close() {
        this.dialogRef.close();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDialogOverlay, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpDialogOverlay, isStandalone: true, selector: "[ngpDialogOverlay]", host: { listeners: { "click": "close()" } }, providers: [{ provide: NgpDialogOverlayToken, useExisting: NgpDialogOverlay }], exportAs: ["ngpDialogOverlay"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDialogOverlay, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpDialogOverlay]',
                    exportAs: 'ngpDialogOverlay',
                    providers: [{ provide: NgpDialogOverlayToken, useExisting: NgpDialogOverlay }],
                }]
        }], propDecorators: { close: [{
                type: HostListener,
                args: ['click']
            }] } });
