// ng-primitives/esm2022/pagination/pagination/pagination.directive.mjs
import { booleanAttribute, computed, Directive, input, model, numberAttribute, } from '@angular/core';
import { NgpPaginationToken } from './pagination.token';
import * as i0 from "@angular/core";
export class NgpPagination {
    constructor() {
        /**
         * The currently selected page.
         */
        this.page = model(1, {
            alias: 'ngpPaginationPage',
        });
        /**
         * The total number of pages.
         */
        this.pageCount = input(0, {
            alias: 'ngpPaginationPageCount',
            transform: numberAttribute,
        });
        /**
         * Whether the pagination is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpPaginationDisabled',
            transform: booleanAttribute,
        });
        /**
         * Determine if we are on the first page.
         * @internal
         */
        this.firstPage = computed(() => this.page() === 1);
        /**
         * Determine if we are on the last page.
         * @internal
         */
        this.lastPage = computed(() => this.page() === this.pageCount());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpPagination, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpPagination, isStandalone: true, selector: "[ngpPagination]", inputs: { page: { classPropertyName: "page", publicName: "ngpPaginationPage", isSignal: true, isRequired: false, transformFunction: null }, pageCount: { classPropertyName: "pageCount", publicName: "ngpPaginationPageCount", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpPaginationDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { page: "ngpPaginationPageChange" }, host: { attributes: { "role": "navigation" }, properties: { "attr.data-page": "page()", "attr.data-page-count": "pageCount()", "attr.data-first-page": "firstPage() ? \"\" : null", "attr.data-last-page": "lastPage() ? \"\" : null", "attr.data-disabled": "disabled() ? \"\" : null" } }, providers: [{ provide: NgpPaginationToken, useExisting: NgpPagination }], exportAs: ["ngpPagination"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpPagination, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpPagination]',
                    exportAs: 'ngpPagination',
                    providers: [{ provide: NgpPaginationToken, useExisting: NgpPagination }],
                    host: {
                        role: 'navigation',
                        '[attr.data-page]': 'page()',
                        '[attr.data-page-count]': 'pageCount()',
                        '[attr.data-first-page]': 'firstPage() ? "" : null',
                        '[attr.data-last-page]': 'lastPage() ? "" : null',
                        '[attr.data-disabled]': 'disabled() ? "" : null',
                    },
                }]
        }] });
