// ng-primitives/esm2022/resize/resize/resize.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { DestroyRef, Directive, ElementRef, NgZone, inject, output } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { fromResizeEvent } from '../utils/resize';
import * as i0 from "@angular/core";
export class NgpResize {
    constructor() {
        /**
         * Access the element.
         */
        this.element = inject(ElementRef);
        /**
         * Access zone.js
         */
        this.ngZone = inject(NgZone);
        /**
         * Access the destroy ref
         */
        this.destroyRef = inject(DestroyRef);
        /**
         * Emit when the element is resized.
         */
        this.didResize = output({
            alias: 'ngpResize',
        });
    }
    ngOnInit() {
        // oberve the element for resize events
        fromResizeEvent(this.element.nativeElement)
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe(event => this.ngZone.run(() => this.didResize.emit(event)));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpResize, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpResize, isStandalone: true, selector: "[ngpResize]", outputs: { didResize: "ngpResize" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpResize, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpResize]',
                    standalone: true,
                }]
        }] });
