// ng-primitives/esm2022/switch/switch-thumb/switch-thumb.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive } from '@angular/core';
import { NgpFocusVisible, NgpHover, NgpPress } from 'ng-primitives/interactions';
import { injectSwitch } from '../switch/switch.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/interactions";
export class NgpSwitchThumb {
    constructor() {
        /**
         * Access the switch directive.
         */
        this.switch = injectSwitch();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSwitchThumb, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpSwitchThumb, isStandalone: true, selector: "[ngpSwitchThumb]", host: { properties: { "attr.data-checked": "switch.checked() ? \"\" : null", "attr.data-disabled": "switch.disabled() ? \"\" : null" } }, hostDirectives: [{ directive: i1.NgpHover }, { directive: i1.NgpFocusVisible }, { directive: i1.NgpPress }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSwitchThumb, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpSwitchThumb]',
                    standalone: true,
                    host: {
                        '[attr.data-checked]': 'switch.checked() ? "" : null',
                        '[attr.data-disabled]': 'switch.disabled() ? "" : null',
                    },
                    hostDirectives: [NgpHover, NgpFocusVisible, NgpPress],
                }]
        }] });
