// ng-primitives/esm2022/roving-focus/roving-focus-item/roving-focus-item.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { FocusMonitor } from '@angular/cdk/a11y';
import { Directive, ElementRef, HostListener, booleanAttribute, computed, inject, input, } from '@angular/core';
import { injectRovingFocusGroup } from '../roving-focus-group/roving-focus-group.token';
import { NgpRovingFocusItemToken } from './roving-focus-item.token';
import * as i0 from "@angular/core";
export class NgpRovingFocusItem {
    constructor() {
        /**
         * Access the group the roving focus item belongs to.
         */
        this.group = injectRovingFocusGroup();
        /**
         * Access the focus monitor service.
         */
        this.focusMonitor = inject(FocusMonitor);
        /**
         * Access the element the roving focus item is attached to.
         */
        this.elementRef = inject(ElementRef);
        /**
         * Define if the item is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpRovingFocusItemDisabled',
            transform: booleanAttribute,
        });
        /**
         * Derive the tabindex of the roving focus item.
         */
        this.tabindex = computed(() => !this.group.disabled() && this.group.activeItem() === this ? 0 : -1);
    }
    /**
     * Initialize the roving focus item.
     */
    ngOnInit() {
        this.group.register(this);
    }
    /**
     * Clean up the roving focus item.
     */
    ngOnDestroy() {
        this.group.unregister(this);
    }
    /**
     * Forward the keydown event to the roving focus group.
     * @param event The keyboard event
     */
    onKeydown(event) {
        if (this.disabled()) {
            return;
        }
        this.group.onKeydown(event);
    }
    /**
     * Activate the roving focus item on click.
     */
    activate() {
        if (this.disabled()) {
            return;
        }
        this.group.setActiveItem(this, 'mouse');
    }
    /**
     * Focus the roving focus item.
     * @param origin The origin of the focus
     */
    focus(origin) {
        this.focusMonitor.focusVia(this.elementRef, origin);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpRovingFocusItem, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpRovingFocusItem, isStandalone: true, selector: "[ngpRovingFocusItem]", inputs: { disabled: { classPropertyName: "disabled", publicName: "ngpRovingFocusItemDisabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { listeners: { "keydown": "onKeydown($event)", "click": "activate()" }, properties: { "attr.tabindex": "tabindex()" } }, providers: [{ provide: NgpRovingFocusItemToken, useExisting: NgpRovingFocusItem }], exportAs: ["ngpRovingFocusItem"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpRovingFocusItem, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpRovingFocusItem]',
                    exportAs: 'ngpRovingFocusItem',
                    providers: [{ provide: NgpRovingFocusItemToken, useExisting: NgpRovingFocusItem }],
                    host: {
                        '[attr.tabindex]': 'tabindex()',
                    },
                }]
        }], propDecorators: { onKeydown: [{
                type: HostListener,
                args: ['keydown', ['$event']]
            }], activate: [{
                type: HostListener,
                args: ['click']
            }] } });
