// ng-primitives/esm2022/tabs/tabset/tabset.directive.mjs
import { Directive, booleanAttribute, computed, contentChildren, input, model, } from '@angular/core';
import { NgpOrientationToken } from 'ng-primitives/internal';
import { NgpRovingFocusGroup } from 'ng-primitives/roving-focus';
import { uniqueId } from 'ng-primitives/utils';
import { injectTabsConfig } from '../config/tabs.config';
import { NgpTabPanelToken } from '../tab-panel/tab-panel.token';
import { NgpTabsetToken } from './tabset.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/roving-focus";
export class NgpTabset {
    constructor() {
        /**
         * Access the global tabset configuration
         */
        this.config = injectTabsConfig();
        /**
         * Define the id for the tabset
         */
        this.id = input(uniqueId('ngp-tabset'));
        /**
         * Define the active tab
         */
        this.value = model(null, {
            alias: 'ngpTabsetValue',
        });
        /**
         * The orientation of the tabset
         * @default 'horizontal'
         */
        this.orientation = input(this.config.orientation, {
            alias: 'ngpTabsetOrientation',
        });
        /**
         * Whether tabs should activate on focus
         */
        this.activateOnFocus = input(this.config.activateOnFocus, {
            alias: 'ngpTabsetActivateOnFocus',
            transform: booleanAttribute,
        });
        /**
         * Access the tabs within the tabset
         */
        this.panels = contentChildren(NgpTabPanelToken, { descendants: true });
        /**
         * Get the id of the selected tab
         */
        this.selectedTab = computed(() => {
            const panels = this.panels();
            // if there is a value set and a tab with that value exists, return the value
            if (panels.some(panel => panel.value() === this.value())) {
                return this.value();
            }
            // otherwise return the first tab
            return panels[0]?.value();
        });
    }
    /**
     * Select a tab by its value
     * @param value The value of the tab to select
     */
    select(value) {
        this.value.set(value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTabset, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.2.0", version: "18.2.13", type: NgpTabset, isStandalone: true, selector: "[ngpTabset]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, value: { classPropertyName: "value", publicName: "ngpTabsetValue", isSignal: true, isRequired: false, transformFunction: null }, orientation: { classPropertyName: "orientation", publicName: "ngpTabsetOrientation", isSignal: true, isRequired: false, transformFunction: null }, activateOnFocus: { classPropertyName: "activateOnFocus", publicName: "ngpTabsetActivateOnFocus", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { value: "ngpTabsetValueChange" }, host: { properties: { "attr.id": "id()", "attr.data-orientation": "orientation()" } }, providers: [
            { provide: NgpTabsetToken, useExisting: NgpTabset },
            { provide: NgpOrientationToken, useExisting: NgpTabset },
        ], queries: [{ propertyName: "panels", predicate: NgpTabPanelToken, descendants: true, isSignal: true }], exportAs: ["ngpTabset"], hostDirectives: [{ directive: i1.NgpRovingFocusGroup }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTabset, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpTabset]',
                    exportAs: 'ngpTabset',
                    providers: [
                        { provide: NgpTabsetToken, useExisting: NgpTabset },
                        { provide: NgpOrientationToken, useExisting: NgpTabset },
                    ],
                    hostDirectives: [NgpRovingFocusGroup],
                    host: {
                        '[attr.id]': 'id()',
                        '[attr.data-orientation]': 'orientation()',
                    },
                }]
        }] });
