// ng-primitives/esm2022/date-picker/date-picker-label/date-picker-label.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, input } from '@angular/core';
import { uniqueId } from 'ng-primitives/utils';
import { injectDatePicker } from '../date-picker/date-picker.token';
import { NgpDatePickerLabelToken } from './date-picker-label.token';
import * as i0 from "@angular/core";
export class NgpDatePickerLabel {
    constructor() {
        /**
         * Access the date picker.
         */
        this.datePicker = injectDatePicker();
        /**
         * Define a unique id for the label.
         */
        this.id = input(uniqueId('ngp-date-picker-label'));
        /**
         * Define the aria live attribute.
         */
        this.ariaLive = input('polite', {
            alias: 'aria-live',
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDatePickerLabel, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpDatePickerLabel, isStandalone: true, selector: "[ngpDatePickerLabel]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, ariaLive: { classPropertyName: "ariaLive", publicName: "aria-live", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "id": "id()", "attr.aria-live": "ariaLive()", "attr.data-disabled": "datePicker.disabled() ? \"\" : null" } }, providers: [{ provide: NgpDatePickerLabelToken, useExisting: NgpDatePickerLabel }], exportAs: ["ngpDatePickerLabel"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDatePickerLabel, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpDatePickerLabel]',
                    exportAs: 'ngpDatePickerLabel',
                    providers: [{ provide: NgpDatePickerLabelToken, useExisting: NgpDatePickerLabel }],
                    host: {
                        '[id]': 'id()',
                        '[attr.aria-live]': 'ariaLive()',
                        '[attr.data-disabled]': 'datePicker.disabled() ? "" : null',
                    },
                }]
        }] });
