// ng-primitives/esm2022/tabs/tab-panel/tab-panel.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, computed, input } from '@angular/core';
import { injectTabset } from '../tabset/tabset.token';
import { NgpTabPanelToken } from './tab-panel.token';
import * as i0 from "@angular/core";
export class NgpTabPanel {
    constructor() {
        /**
         * Access the tabset
         */
        this.tabset = injectTabset();
        /**
         * The value of the tab
         */
        this.value = input.required({ alias: 'ngpTabPanelValue' });
        /**
         * Determine the id of the tab panel
         * @internal
         */
        this.id = input();
        /**
         * Determine a unique id for the tab panel if not provided
         * @internal
         */
        this.defaultId = computed(() => `${this.tabset.id()}-panel-${this.value()}`);
        /**
         * Determine the aria-labelledby of the tab panel
         * @internal
         */
        this.labelledBy = computed(() => `${this.tabset.id()}-button-${this.value()}`);
        /**
         * Whether the tab is active
         */
        this.active = computed(() => this.tabset.selectedTab() === this.value());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTabPanel, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpTabPanel, isStandalone: true, selector: "[ngpTabPanel]", inputs: { value: { classPropertyName: "value", publicName: "ngpTabPanelValue", isSignal: true, isRequired: true, transformFunction: null }, id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "tabpanel", "tabIndex": "0" }, properties: { "attr.aria-labelledby": "labelledBy()", "attr.data-active": "active() ? \"\" : null", "attr.data-orientation": "tabset.orientation()" } }, providers: [{ provide: NgpTabPanelToken, useExisting: NgpTabPanel }], exportAs: ["ngpTabPanel"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTabPanel, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpTabPanel]',
                    exportAs: 'ngpTabPanel',
                    providers: [{ provide: NgpTabPanelToken, useExisting: NgpTabPanel }],
                    host: {
                        role: 'tabpanel',
                        tabIndex: '0',
                        '[attr.aria-labelledby]': 'labelledBy()',
                        '[attr.data-active]': 'active() ? "" : null',
                        '[attr.data-orientation]': 'tabset.orientation()',
                    },
                }]
        }] });
