// ng-primitives/esm2022/menu/menu-item/menu-item.directive.mjs
import { CdkMenuItem } from '@angular/cdk/menu';
import { booleanAttribute, Directive, input } from '@angular/core';
import { NgpFocusVisible, NgpHover, NgpPress } from 'ng-primitives/interactions';
import { NgpMenuItemToken } from './menu-item.token';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/menu";
import * as i2 from "ng-primitives/interactions";
export class NgpMenuItem {
    constructor() {
        /**
         * Whether the menu item is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpMenuItemDisabled',
            transform: booleanAttribute,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpMenuItem, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpMenuItem, isStandalone: true, selector: "[ngpMenuItem]", inputs: { disabled: { classPropertyName: "disabled", publicName: "ngpMenuItemDisabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "attr.data-disabled": "disabled() ? \"\" : null" } }, providers: [{ provide: NgpMenuItemToken, useExisting: NgpMenuItem }], exportAs: ["ngpMenuItem"], hostDirectives: [{ directive: i1.CdkMenuItem, inputs: ["cdkMenuItemDisabled", "ngpMenuItemDisabled"] }, { directive: i2.NgpFocusVisible }, { directive: i2.NgpHover }, { directive: i2.NgpPress }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpMenuItem, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpMenuItem]',
                    exportAs: 'ngpMenuItem',
                    providers: [{ provide: NgpMenuItemToken, useExisting: NgpMenuItem }],
                    hostDirectives: [
                        { directive: CdkMenuItem, inputs: ['cdkMenuItemDisabled:ngpMenuItemDisabled'] },
                        NgpFocusVisible,
                        NgpHover,
                        NgpPress,
                    ],
                    host: {
                        '[attr.data-disabled]': 'disabled() ? "" : null',
                    },
                }]
        }] });
