// ng-primitives/esm2022/radio/radio-item/radio-item.directive.mjs
import { Directive, HostListener, booleanAttribute, computed, input } from '@angular/core';
import { NgpFocusVisible, NgpHover, NgpPress } from 'ng-primitives/interactions';
import { NgpRovingFocusItem } from 'ng-primitives/roving-focus';
import { injectRadioGroup } from '../radio-group/radio-group.token';
import { NgpRadioItemToken } from './radio-item.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/roving-focus";
import * as i2 from "ng-primitives/interactions";
export class NgpRadioItem {
    constructor() {
        /**
         * Access the radio group.
         */
        this.radioGroup = injectRadioGroup();
        /**
         * The value of the radio item.
         */
        this.value = input.required({ alias: 'ngpRadioItemValue' });
        /**
         * Whether the radio item is disabled.
         * @default false
         */
        this.disabled = input(false, {
            alias: 'ngpRadioItemDisabled',
            transform: booleanAttribute,
        });
        /**
         * Whether the radio item is checked.
         */
        this.checked = computed(() => this.radioGroup.value() === this.value());
    }
    /**
     * When the item receives focus, select it.
     * @internal
     */
    onFocus() {
        this.radioGroup.select(this.value());
    }
    /**
     * When the item receives a click, select it.
     * @internal
     */
    onClick() {
        this.radioGroup.select(this.value());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpRadioItem, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpRadioItem, isStandalone: true, selector: "[ngpRadioItem]", inputs: { value: { classPropertyName: "value", publicName: "ngpRadioItemValue", isSignal: true, isRequired: true, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpRadioItemDisabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "radio" }, listeners: { "focus": "onFocus()", "click": "onClick()" }, properties: { "attr.aria-checked": "checked() ? \"true\" : \"false\"", "attr.data-disabled": "disabled() ? \"\" : null", "attr.data-checked": "checked() ? \"\" : null" } }, providers: [{ provide: NgpRadioItemToken, useExisting: NgpRadioItem }], hostDirectives: [{ directive: i1.NgpRovingFocusItem }, { directive: i2.NgpHover }, { directive: i2.NgpFocusVisible }, { directive: i2.NgpPress }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpRadioItem, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpRadioItem]',
                    standalone: true,
                    hostDirectives: [NgpRovingFocusItem, NgpHover, NgpFocusVisible, NgpPress],
                    providers: [{ provide: NgpRadioItemToken, useExisting: NgpRadioItem }],
                    host: {
                        role: 'radio',
                        '[attr.aria-checked]': 'checked() ? "true" : "false"',
                        '[attr.data-disabled]': 'disabled() ? "" : null',
                        '[attr.data-checked]': 'checked() ? "" : null',
                    },
                }]
        }], propDecorators: { onFocus: [{
                type: HostListener,
                args: ['focus']
            }], onClick: [{
                type: HostListener,
                args: ['click']
            }] } });
