// ng-primitives/esm2022/progress/progress-indicator/progress-indicator.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive } from '@angular/core';
import { injectProgress } from '../progress/progress.token';
import * as i0 from "@angular/core";
export class NgpProgressIndicator {
    constructor() {
        /**
         * Access the progress directive.
         */
        this.progress = injectProgress();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpProgressIndicator, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpProgressIndicator, isStandalone: true, selector: "[ngpProgressIndicator]", host: { properties: { "attr.data-state": "progress.state()", "attr.data-value": "progress.value()", "attr.data-max": "progress.max()" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpProgressIndicator, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpProgressIndicator]',
                    standalone: true,
                    host: {
                        '[attr.data-state]': 'progress.state()',
                        '[attr.data-value]': 'progress.value()',
                        '[attr.data-max]': 'progress.max()',
                    },
                }]
        }] });
