// ng-primitives/esm2022/toast/toast/toast-ref.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
export class NgpToastRef {
    /** Get the toast height */
    get height() {
        return this.toastElement.offsetHeight;
    }
    constructor(
    /** Store the toast element */
    toastElement, 
    /** Store the duration */
    duration, 
    /** The position of the toast */
    position, 
    /** The gravity of the toast */
    gravity, 
    /** Whether we should stop on focus */
    stopOnHover, 
    /** The aria live setting */
    ariaLive, onDismiss) {
        this.toastElement = toastElement;
        this.duration = duration;
        this.position = position;
        this.gravity = gravity;
        this.stopOnHover = stopOnHover;
        this.ariaLive = ariaLive;
        this.onDismiss = onDismiss;
        /** Store the current timeout */
        this.timeoutId = null;
        this.toastElement.setAttribute('data-toast', 'visible');
        this.setPosition(position);
        this.setGravity(gravity);
        this.setAriaLive(ariaLive);
        this.setupTimeouts();
        this.setupListeners();
    }
    dismiss() {
        // determine if there is a transition on the element
        const transitionDuration = parseFloat(getComputedStyle(this.toastElement).transitionDuration);
        // if there is no transition, dismiss immediately
        if (transitionDuration === 0) {
            this.removeElement();
            return;
        }
        // wait for the transition to end
        this.toastElement.addEventListener('transitionend', () => this.removeElement());
        this.toastElement.setAttribute('data-toast', 'hidden');
    }
    removeElement() {
        this.toastElement.parentNode?.removeChild(this.toastElement);
        this.onDismiss();
    }
    /** Setup duration timeouts */
    setupTimeouts() {
        // if the duration is 0 skip
        if (this.duration === 0) {
            return;
        }
        this.timeoutId = window.setTimeout(() => this.dismiss(), this.duration);
    }
    setupListeners() {
        if (!this.stopOnHover) {
            return;
        }
        // setup event listeners if we should stop on focus
        this.toastElement.addEventListener('mouseover', () => {
            window.clearTimeout(this.timeoutId);
            this.timeoutId = null;
        });
        this.toastElement.addEventListener('mouseleave', () => this.setupTimeouts());
    }
    /** Set the position attribute */
    setPosition(position) {
        this.toastElement.setAttribute('data-position', position);
    }
    /** Set the gravity attribute */
    setGravity(gravity) {
        this.toastElement.setAttribute('data-gravity', gravity);
    }
    /** Set the aria live attribute */
    setAriaLive(ariaLive) {
        this.toastElement.setAttribute('aria-live', ariaLive);
    }
    /**
     * @internal
     */
    setInset(property, value) {
        this.toastElement.style[property] = value;
    }
}
