// ng-primitives/esm2022/date-picker/date-picker-grid/date-picker-grid.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { computed, Directive } from '@angular/core';
import { injectDatePicker } from '../date-picker/date-picker.token';
import { NgpDatePickerGridToken } from './date-picker-grid.token';
import * as i0 from "@angular/core";
export class NgpDatePickerGrid {
    constructor() {
        /**
         * Access the date picker.
         */
        this.datePicker = injectDatePicker();
        /**
         * Determine the id for the label.
         */
        this.labelId = computed(() => this.datePicker.label()?.id());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDatePickerGrid, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpDatePickerGrid, isStandalone: true, selector: "[ngpDatePickerGrid]", host: { attributes: { "role": "grid" }, properties: { "attr.aria-labelledby": "labelId()", "attr.data-disabled": "datePicker.disabled() ? \"\" : null" } }, providers: [{ provide: NgpDatePickerGridToken, useExisting: NgpDatePickerGrid }], exportAs: ["ngpDatePickerGrid"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDatePickerGrid, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpDatePickerGrid]',
                    exportAs: 'ngpDatePickerGrid',
                    providers: [{ provide: NgpDatePickerGridToken, useExisting: NgpDatePickerGrid }],
                    host: {
                        role: 'grid',
                        '[attr.aria-labelledby]': 'labelId()',
                        '[attr.data-disabled]': 'datePicker.disabled() ? "" : null',
                    },
                }]
        }] });
