// ng-primitives/esm2022/accordion/accordion/accordion.directive.mjs
import { Directive, booleanAttribute, input, model } from '@angular/core';
import { injectAccordionConfig } from '../config/accordion.config';
import { NgpAccordionToken } from './accordion.token';
import * as i0 from "@angular/core";
export class NgpAccordion {
    constructor() {
        /**
         * Access the global accordion configuration.
         */
        this.config = injectAccordionConfig();
        /**
         * The type of the accordion.
         */
        this.type = input(this.config.type, {
            alias: 'ngpAccordionType',
        });
        /**
         * Whether the accordion is collapsible.
         */
        this.collapsible = input(this.config.collapsible, {
            alias: 'ngpAccordionCollapsible',
            transform: booleanAttribute,
        });
        /**
         * The value of the accordion.
         */
        this.value = model(null, {
            alias: 'ngpAccordionValue',
        });
        /**
         * Whether the accordion is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpAccordionDisabled',
            transform: booleanAttribute,
        });
        /**
         * The accordion orientation.
         */
        this.orientation = input(this.config.orientation, {
            alias: 'ngpAccordionOrientation',
        });
    }
    /**
     * @param value The value to check.
     * @returns Whether the value is open.
     * @internal
     */
    isOpen(value) {
        if (this.type() === 'multiple') {
            return this.value()?.includes(value) ?? false;
        }
        return this.value() === value;
    }
    toggle(value) {
        const isOpen = this.isOpen(value);
        // if we are in single mode and the value is already open and the accordion is not collapsible, do nothing
        if (this.type() === 'single' && isOpen && !this.collapsible()) {
            return;
        }
        // if we are in single mode then toggle the value
        if (this.type() === 'single') {
            this.value.set(isOpen ? null : value);
            return;
        }
        // if we are in multiple mode then toggle the value
        const values = this.value() ?? [];
        if (isOpen) {
            this.value.set(values.filter(v => v !== value));
        }
        else {
            this.value.set([...values, value]);
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAccordion, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpAccordion, isStandalone: true, selector: "[ngpAccordion]", inputs: { type: { classPropertyName: "type", publicName: "ngpAccordionType", isSignal: true, isRequired: false, transformFunction: null }, collapsible: { classPropertyName: "collapsible", publicName: "ngpAccordionCollapsible", isSignal: true, isRequired: false, transformFunction: null }, value: { classPropertyName: "value", publicName: "ngpAccordionValue", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpAccordionDisabled", isSignal: true, isRequired: false, transformFunction: null }, orientation: { classPropertyName: "orientation", publicName: "ngpAccordionOrientation", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { value: "ngpAccordionValueChange" }, host: { properties: { "attr.data-orientation": "orientation()", "attr.data-disabled": "disabled() ? \"\" : null" } }, providers: [{ provide: NgpAccordionToken, useExisting: NgpAccordion }], exportAs: ["ngpAccordion"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAccordion, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpAccordion]',
                    exportAs: 'ngpAccordion',
                    providers: [{ provide: NgpAccordionToken, useExisting: NgpAccordion }],
                    host: {
                        '[attr.data-orientation]': 'orientation()',
                        '[attr.data-disabled]': 'disabled() ? "" : null',
                    },
                }]
        }] });
