// ng-primitives/esm2022/form-field/form-control/form-control.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, computed, effect, input } from '@angular/core';
import { injectDisabled } from 'ng-primitives/internal';
import { uniqueId } from 'ng-primitives/utils';
import { injectFormField } from '../form-field/form-field.token';
import { NgpFormControlToken } from './form-control.token';
import * as i0 from "@angular/core";
export class NgpFormControl {
    constructor() {
        /**
         * Access the form field that the form control is associated with.
         */
        this.formField = injectFormField();
        /**
         * Whether the form control is disabled by a parent.
         */
        this.disabled = injectDisabled();
        /**
         * The id of the form control. If not provided, a unique id will be generated.
         */
        this.id = input(uniqueId('ngp-form-control'));
        /**
         * Determine the aria-labelledby attribute value.
         */
        this.ariaLabelledBy = computed(() => this.formField?.labels().join(' '));
        /**
         * Determine the aria-describedby attribute value.
         */
        this.ariaDescribedBy = computed(() => this.formField?.descriptions().join(' '));
        effect(onCleanup => {
            this.formField?.setFormControl(this.id());
            onCleanup(() => this.formField?.removeFormControl());
        }, { allowSignalWrites: true });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpFormControl, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpFormControl, isStandalone: true, selector: "[ngpFormControl]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "id": "id()", "attr.aria-labelledby": "ariaLabelledBy()", "attr.aria-describedby": "ariaDescribedBy()", "attr.data-invalid": "formField?.invalid() ? \"\" : null", "attr.data-valid": "formField?.valid() ? \"\" : null", "attr.data-touched": "formField?.touched() ? \"\" : null", "attr.data-pristine": "formField?.pristine() ? \"\" : null", "attr.data-dirty": "formField?.dirty() ? \"\" : null", "attr.data-pending": "formField?.pending() ? \"\" : null", "attr.data-disabled": "formField?.disabled() || disabled() ? \"\" : null" } }, providers: [{ provide: NgpFormControlToken, useExisting: NgpFormControl }], exportAs: ["ngpFormControl"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpFormControl, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpFormControl]',
                    exportAs: 'ngpFormControl',
                    providers: [{ provide: NgpFormControlToken, useExisting: NgpFormControl }],
                    host: {
                        '[id]': 'id()',
                        '[attr.aria-labelledby]': 'ariaLabelledBy()',
                        '[attr.aria-describedby]': 'ariaDescribedBy()',
                        '[attr.data-invalid]': 'formField?.invalid() ? "" : null',
                        '[attr.data-valid]': 'formField?.valid() ? "" : null',
                        '[attr.data-touched]': 'formField?.touched() ? "" : null',
                        '[attr.data-pristine]': 'formField?.pristine() ? "" : null',
                        '[attr.data-dirty]': 'formField?.dirty() ? "" : null',
                        '[attr.data-pending]': 'formField?.pending() ? "" : null',
                        '[attr.data-disabled]': 'formField?.disabled() || disabled() ? "" : null',
                    },
                }]
        }], ctorParameters: () => [] });
