// @ngx-chart/doughnut/esm2015/lib/doughnut-graph.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/doughnut-graph.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef } from '@angular/core';
import * as d3 from 'd3';
import { WrapperComponentD3, CircularMeasures } from '@ngx-chart/common';
import { DoughnutGraph } from './doughnut.class';
export class DoughnutGraphComponent extends DoughnutGraph {
    /**
     * @param {?} elementRef
     * @param {?} circularMeasures
     */
    constructor(elementRef, circularMeasures) {
        super();
        this.elementRef = elementRef;
        this.circularMeasures = circularMeasures;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        // Remove previous graphs
        if (changes.graphData.previousValue) {
            d3.select(`#svg-${this.currentElementId}`).remove();
        }
        this.graphElement = new WrapperComponentD3(this.elementRef);
        this.currentElementId = this.graphElement.uid;
        this.minSize = this.circularMeasures.calcRadius(this.graphElement.width, this.graphElement.height);
        // If graph has labels and the position of if are extenal reduce radius in 25 points.
        this.minSize = this.graphData.labels && this.graphData.labels.position === 'external'
            ? this.minSize - 25
            : this.minSize;
        this.segmentsDegree = this.scaleData(this.graphData.values).map((/**
         * @param {?} value
         * @return {?}
         */
        value => this.circularMeasures.arcValue(value)));
        this.graphData.colors = this.checkGraphDataColors();
        this.segmentsDegree.map((/**
         * @param {?} arc
         * @param {?} idx
         * @return {?}
         */
        (arc, idx) => {
            /** @type {?} */
            const startAngle = idx === 0
                ? 0
                : this.segmentsDegree.slice(0, idx).reduce((/**
                 * @param {?} curr
                 * @param {?} prev
                 * @return {?}
                 */
                (curr, prev) => curr + prev));
            /** @type {?} */
            const endAngle = idx === this.segmentsDegree.length
                ? 360
                : this.segmentsDegree.slice(0, idx + 1).reduce((/**
                 * @param {?} curr
                 * @param {?} prev
                 * @return {?}
                 */
                (curr, prev) => curr + prev));
            this.drawSegment(startAngle, endAngle, idx);
        }));
    }
}
DoughnutGraphComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngl-doughnut',
                template: ``,
                providers: [
                    CircularMeasures
                ]
            }] }
];
/** @nocollapse */
DoughnutGraphComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: CircularMeasures }
];
if (false) {
    /** @type {?} */
    DoughnutGraphComponent.prototype.currentElementId;
    /** @type {?} */
    DoughnutGraphComponent.prototype.elementRef;
    /** @type {?} */
    DoughnutGraphComponent.prototype.circularMeasures;
}
