// ng-primitives/esm2022/slider/slider-range/slider-range.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive } from '@angular/core';
import { injectSlider } from '../slider/slider.token';
import { NgpSliderRangeToken } from './slider-range.token';
import * as i0 from "@angular/core";
export class NgpSliderRange {
    constructor() {
        /**
         * Access the slider.
         */
        this.slider = injectSlider();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSliderRange, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpSliderRange, isStandalone: true, selector: "[ngpSliderRange]", host: { properties: { "attr.data-orientation": "slider.orientation()", "attr.data-disabled": "slider.disabled() ? \"\" : null", "style.width.%": "slider.orientation() === \"horizontal\" ? slider.percentage() : undefined", "style.height.%": "slider.orientation() === \"vertical\" ? slider.percentage() : undefined" } }, providers: [{ provide: NgpSliderRangeToken, useExisting: NgpSliderRange }], exportAs: ["ngpSliderRange"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSliderRange, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpSliderRange]',
                    exportAs: 'ngpSliderRange',
                    providers: [{ provide: NgpSliderRangeToken, useExisting: NgpSliderRange }],
                    host: {
                        '[attr.data-orientation]': 'slider.orientation()',
                        '[attr.data-disabled]': 'slider.disabled() ? "" : null',
                        '[style.width.%]': 'slider.orientation() === "horizontal" ? slider.percentage() : undefined',
                        '[style.height.%]': 'slider.orientation() === "vertical" ? slider.percentage() : undefined',
                    },
                }]
        }] });
