// ng-primitives/fesm2022/ng-primitives-a11y.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, signal, computed, Directive } from '@angular/core';

/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
const NgpVisuallyHiddenToken = new InjectionToken('NgpVisuallyHiddenToken');
/**
 * Inject the VisuallyHidden directive instance
 */
function injectVisuallyHidden() {
    return inject(NgpVisuallyHiddenToken);
}

/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
class NgpVisuallyHidden {
    constructor() {
        /**
         * Whether the element is hidden.
         */
        this.hidden = signal(true);
        this.style = computed(() => {
            if (!this.hidden()) {
                return {};
            }
            return {
                position: 'absolute',
                width: '1px',
                height: '1px',
                margin: '-1px',
                padding: '0',
                overflow: 'hidden',
                clip: 'rect(0, 0, 0, 0)',
                whiteSpace: 'nowrap',
                border: '0',
                wordWrap: 'normal',
                outline: '0',
                '-webkit-appearance': 'none',
                '-moz-appearance': 'none',
                'inset-inline-start': '0',
            };
        });
    }
    /**
     * Set the element visibility.
     * @param visible
     */
    setVisibility(visible) {
        this.hidden.set(!visible);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpVisuallyHidden, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpVisuallyHidden, isStandalone: true, selector: "[ngpVisuallyHidden]", host: { properties: { "style": "style()" } }, providers: [{ provide: NgpVisuallyHiddenToken, useExisting: NgpVisuallyHidden }], exportAs: ["ngpVisuallyHidden"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpVisuallyHidden, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpVisuallyHidden]',
                    exportAs: 'ngpVisuallyHidden',
                    providers: [{ provide: NgpVisuallyHiddenToken, useExisting: NgpVisuallyHidden }],
                    host: {
                        '[style]': 'style()',
                    },
                }]
        }] });

/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Generated bundle index. Do not edit.
 */

export { NgpVisuallyHidden, NgpVisuallyHiddenToken, injectVisuallyHidden };
