// ng-primitives/fesm2022/ng-primitives-textarea.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, booleanAttribute, Directive } from '@angular/core';
import * as i1 from 'ng-primitives/form-field';
import { NgpFormControl } from 'ng-primitives/form-field';
import * as i2 from 'ng-primitives/interactions';
import { NgpHover, NgpFocus, NgpPress } from 'ng-primitives/interactions';
import { NgpDisabledToken } from 'ng-primitives/internal';

/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
const NgpTextareaToken = new InjectionToken('NgpTextareaToken');
/**
 * Inject the Textarea directive instance
 */
function injectTextarea() {
    return inject(NgpTextareaToken);
}

class NgpTextarea {
    constructor() {
        /**
         * Whether the element is disabled.
         */
        this.disabled = input(false, {
            transform: booleanAttribute,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTextarea, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpTextarea, isStandalone: true, selector: "[ngpTextarea]", inputs: { disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null } }, providers: [
            { provide: NgpTextareaToken, useExisting: NgpTextarea },
            { provide: NgpDisabledToken, useExisting: NgpTextarea },
        ], exportAs: ["ngpTextarea"], hostDirectives: [{ directive: i1.NgpFormControl }, { directive: i2.NgpHover }, { directive: i2.NgpFocus }, { directive: i2.NgpPress }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTextarea, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpTextarea]',
                    exportAs: 'ngpTextarea',
                    providers: [
                        { provide: NgpTextareaToken, useExisting: NgpTextarea },
                        { provide: NgpDisabledToken, useExisting: NgpTextarea },
                    ],
                    hostDirectives: [NgpFormControl, NgpHover, NgpFocus, NgpPress],
                }]
        }] });

/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Generated bundle index. Do not edit.
 */

export { NgpTextarea, NgpTextareaToken };
