// @ngx-chart/common/fesm2015/ngx-chart-common.js
import { select, scaleLinear } from 'd3';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/wrapper-component-d3.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class WrapperComponentD3 {
    /**
     * @param {?} _elementRef
     */
    constructor(_elementRef) {
        // Initialize private variables
        this._uid = this.generateUID();
        this._height = _elementRef.nativeElement.clientHeight;
        this._width = _elementRef.nativeElement.clientWidth;
        // Add Id to Element;
        _elementRef.nativeElement.id = this._uid;
        // Create svg container
        this._svgContainer = select(`#${this._uid}`)
            .append('svg')
            .attr('id', `svg-${this._uid}`)
            .attr('height', this._height)
            .attr('width', this._width);
    }
    /**
     * svgContainer
     * return svg container
     * @return {?} HTMLElement
     */
    get svgContainer() {
        return this._svgContainer;
    }
    /**
     * uid
     * return uid of container
     * @return {?} string
     */
    get uid() {
        return this._uid;
    }
    /**
     * height
     * return height of container
     * @return {?} number
     */
    get height() {
        return this._height;
    }
    /**
     * width
     * return width of container
     * @return {?} number
     */
    get width() {
        return this._width;
    }
    /**
     * generateUID
     * Generate new UID. The uid is make with Date.now and plus random number and then replace it to string.
     * @private
     * @return {?} string
     */
    generateUID() {
        return (Date.now() + Math.random()).toString(36).replace('.', '-');
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/colors.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class Colors {
    constructor() { }
    /**
     * generateRandomPallete
     *
     * Generate random palette of colors
     * @param {?} length
     * @return {?} Array
     */
    generateRandomPallete(length) {
        /** @type {?} */
        const letters = '0123456789ABCDEF';
        /** @type {?} */
        const pallete = new Array(length).fill('');
        return pallete.map((/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            let color = '#';
            for (let i = 0; i < 6; i++) {
                color += letters[Math.floor(Math.random() * 16)];
            }
            return color;
        }));
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/circular-measures.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class CircularMeasures {
    constructor() { }
    /**
     * calcRadius
     *
     * Calculate the minimum size between the height and width of the container element
     * with this value i will calculate the radius of circle.
     *
     * @param {?} width width of container
     * @param {?} height width of container
     * @return {?} number
     */
    calcRadius(width, height) {
        return width > height
            ? height
            : width;
    }
    /**
     * arcValue
     *
     * Calculate the arc to show.
     * Convert the pass value into a number that represent the degrees in a circle
     *
     * @param {?} value value to represent
     * @return {?} number
     */
    arcValue(value) {
        /** @type {?} */
        const interValue = scaleLinear()
            .domain([0, 100])
            .range([0, 360]);
        return interValue(value);
    }
    /**
     * toRadians
     *
     * Convert degrees values in radians
     *
     * @param {?} degree value of degree
     * @return {?} number
     */
    toRadians(degree) {
        return degree * Math.PI / 180;
    }
}

/**
 * @fileoverview added by tsickle
 * Generated from: public_api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: ngx-chart-common.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { WrapperComponentD3, Colors, CircularMeasures };
