// ng-primitives/esm2022/popover/config/popover.config.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { InjectionToken, inject } from '@angular/core';
export const defaultPopoverConfig = {
    offset: 4,
    placement: 'bottom',
    showDelay: 0,
    hideDelay: 0,
    flip: true,
    container: null,
    closeOnOutsideClick: true,
};
export const NgpPopoverConfigToken = new InjectionToken('NgpPopoverConfigToken');
/**
 * Provide the default Popover configuration
 * @param config The Popover configuration
 * @returns The provider
 */
export function providePopoverConfig(config) {
    return [
        {
            provide: NgpPopoverConfigToken,
            useValue: { ...defaultPopoverConfig, ...config },
        },
    ];
}
/**
 * Inject the Popover configuration
 * @returns The global Popover configuration
 */
export function injectPopoverConfig() {
    return inject(NgpPopoverConfigToken, { optional: true }) ?? defaultPopoverConfig;
}
