// @ng-matero/extensions/esm2022/dialog/dialog-module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatDialogModule } from '@angular/material/dialog';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { MtxPipesModule } from '@ng-matero/extensions/core';
import { MtxDialogContainer } from './dialog-container';
import { MtxDialog } from './dialog';
import * as i0 from "@angular/core";
export class MtxDialogModule {
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.0", ngImport: i0, type: MtxDialogModule, imports: [CommonModule,
            MatDialogModule,
            MatButtonModule,
            MatIconModule,
            MtxPipesModule,
            MtxDialogContainer], exports: [MtxDialogContainer] }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxDialogModule, providers: [MtxDialog], imports: [CommonModule,
            MatDialogModule,
            MatButtonModule,
            MatIconModule,
            MtxPipesModule,
            MtxDialogContainer] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatDialogModule,
                        MatButtonModule,
                        MatIconModule,
                        MtxPipesModule,
                        MtxDialogContainer,
                    ],
                    exports: [MtxDialogContainer],
                    providers: [MtxDialog],
                }]
        }] });
