// @ngx-chart/doughnut/esm5/lib/doughnut.class.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/doughnut.class.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import * as d3 from 'd3';
import { Colors } from '@ngx-chart/common';
import { CircularGraphs } from './circular-graphs.class';
/**
 * @abstract
 */
var DoughnutGraph = /** @class */ (function (_super) {
    tslib_1.__extends(DoughnutGraph, _super);
    function DoughnutGraph() {
        return _super.call(this) || this;
    }
    /**
     * scaleData
     *
     * Convert the data values to array of values in percentage scale
     * @param data array of values
     * @returns Array
     */
    /**
     * scaleData
     *
     * Convert the data values to array of values in percentage scale
     * @param {?} data array of values
     * @return {?} Array
     */
    DoughnutGraph.prototype.scaleData = /**
     * scaleData
     *
     * Convert the data values to array of values in percentage scale
     * @param {?} data array of values
     * @return {?} Array
     */
    function (data) {
        /** @type {?} */
        var sumData = data.reduce((/**
         * @param {?} prev
         * @param {?} curr
         * @return {?}
         */
        function (prev, curr) {
            return prev + curr;
        }));
        /** @type {?} */
        var interValue = d3.scaleLinear()
            .domain([0, sumData])
            .range([0, 100]);
        return data.map((/**
         * @param {?} value
         * @return {?}
         */
        function (value) { return interValue(value); }));
    };
    /**
     * checkGraphDataColors
     *
     *
     * Check if the number of colors is the same in the values. If it is different, it generates a random color palette
     * @returns Doughnut['colors']
     */
    /**
     * checkGraphDataColors
     *
     *
     * Check if the number of colors is the same in the values. If it is different, it generates a random color palette
     * @return {?} Doughnut['colors']
     */
    DoughnutGraph.prototype.checkGraphDataColors = /**
     * checkGraphDataColors
     *
     *
     * Check if the number of colors is the same in the values. If it is different, it generates a random color palette
     * @return {?} Doughnut['colors']
     */
    function () {
        /** @type {?} */
        var colors = new Colors();
        return this.graphData.colors
            ? this.graphData.colors.length === this.graphData.values.length
                ? this.graphData.colors
                : tslib_1.__spread(this.graphData.colors, colors.generateRandomPallete(this.graphData.values.length - this.graphData.colors.length || 0))
            : colors.generateRandomPallete(this.graphData.values.length);
    };
    /**
     * drawSegment
     *
     * Draw circular segment
     * @param startAngle start angle of arc
     * @param endAngle end angle of arc
     * @param idx index of arc
     */
    /**
     * drawSegment
     *
     * Draw circular segment
     * @param {?} startAngle start angle of arc
     * @param {?} endAngle end angle of arc
     * @param {?} idx index of arc
     * @return {?}
     */
    DoughnutGraph.prototype.drawSegment = /**
     * drawSegment
     *
     * Draw circular segment
     * @param {?} startAngle start angle of arc
     * @param {?} endAngle end angle of arc
     * @param {?} idx index of arc
     * @return {?}
     */
    function (startAngle, endAngle, idx) {
        /** @type {?} */
        var currenData = {
            id: this.graphElement.uid + "__" + idx,
            radius: this.minSize / 2,
            startAngle: startAngle,
            endAngle: endAngle,
            color: this.graphData.colors[idx],
            padding: this.graphData.padding
        };
        /** @type {?} */
        var arc = this.drawCircle(currenData, this.graphElement);
        if (this.graphData.labels && this.graphData.labels.position === 'internal') {
            this.appendIntLabels(this.graphElement.uid + "__" + idx, arc.centroid(), idx);
        }
        if (this.graphData.labels && this.graphData.labels.position === 'external') {
            this.appendExtLabels(currenData, this.graphElement.uid + "__" + idx, arc.centroid(), idx);
        }
    };
    /**
     * appendExtLabels
     *
     * Append external labels
     * @param arcId arc id
     * @param centro center of arc
     * @param idx index of arc
     */
    /**
     * appendExtLabels
     *
     * Append external labels
     * @param {?} arcId arc id
     * @param {?} centro center of arc
     * @param {?} idx index of arc
     * @return {?}
     */
    DoughnutGraph.prototype.appendIntLabels = /**
     * appendExtLabels
     *
     * Append external labels
     * @param {?} arcId arc id
     * @param {?} centro center of arc
     * @param {?} idx index of arc
     * @return {?}
     */
    function (arcId, centro, idx) {
        this.graphElement.svgContainer
            .select("#" + arcId)
            .append('text')
            .attr('x', centro[0])
            .attr('y', centro[1])
            .attr('dy', '0.33em')
            .attr('dx', '-0.33em')
            .style('fill', 'white')
            .text(this.graphData.labels.titles[idx]);
    };
    /**
     * appendExtLabels
     *
     * Append external labels
     * @param arcId arc id
     * @param centro center of arc
     * @param idx index of arc
     */
    /**
     * appendExtLabels
     *
     * Append external labels
     * @param {?} data
     * @param {?} arcId arc id
     * @param {?} centro center of arc
     * @param {?} idx index of arc
     * @return {?}
     */
    DoughnutGraph.prototype.appendExtLabels = /**
     * appendExtLabels
     *
     * Append external labels
     * @param {?} data
     * @param {?} arcId arc id
     * @param {?} centro center of arc
     * @param {?} idx index of arc
     * @return {?}
     */
    function (data, arcId, centro, idx) {
        /**
         * calculateExternalPos
         *
         * Calculate the position of external pos to draw diagonal line.
         * @param {?} dataArc data of arc
         * @return {?} Array
         */
        function calculateEdgePos(dataArc) {
            /** @type {?} */
            var edgeCircle = d3.arc()
                .innerRadius(dataArc.radius)
                .outerRadius(dataArc.radius)
                .startAngle(this.circularMeasures.toRadians(dataArc.startAngle))
                .endAngle(this.circularMeasures.toRadians(dataArc.endAngle));
            return edgeCircle.centroid();
        }
        /**
         * calculateExternalPos
         *
         * Calculate the position of external pos to draw diagonal line.
         * @param {?} dataArc data of arc
         * @return {?} Array
         */
        function calculateExternalPos(dataArc) {
            /** @type {?} */
            var extCircle = d3.arc()
                .innerRadius(dataArc.radius)
                .outerRadius(dataArc.radius + 20)
                .startAngle(this.circularMeasures.toRadians(dataArc.startAngle))
                .endAngle(this.circularMeasures.toRadians(dataArc.endAngle));
            return extCircle.centroid();
        }
        /** @type {?} */
        var edgeCentro = calculateEdgePos.call(this, data);
        /** @type {?} */
        var extCentro = calculateExternalPos.call(this, data);
        /** @type {?} */
        var paddingHorizontalLabel = extCentro[0] >= 0 ? extCentro[0] + 4 : extCentro[0] - 4;
        /** @type {?} */
        var paddingVerticalLabel = extCentro[1] + 4;
        // Draw line
        this.graphElement.svgContainer
            .select("#" + arcId)
            .append('line')
            .style('stroke', 'black')
            .style('opacity', .7)
            .attr('x1', edgeCentro[0])
            .attr('y1', edgeCentro[1])
            .attr('x2', extCentro[0])
            .attr('y2', extCentro[1]);
        this.graphElement.svgContainer
            .select("#" + arcId)
            .append('text')
            .attr('text-anchor', (/**
         * @return {?}
         */
        function () { return extCentro[0] > 0 ? 'start' : 'end'; }))
            .attr('transform', "translate(" + paddingHorizontalLabel + ", " + paddingVerticalLabel + ")")
            .text(this.graphData.labels.titles[idx]);
    };
    DoughnutGraph.propDecorators = {
        graphData: [{ type: Input }]
    };
    return DoughnutGraph;
}(CircularGraphs));
export { DoughnutGraph };
if (false) {
    /** @type {?} */
    DoughnutGraph.prototype.graphData;
    /**
     * @type {?}
     * @protected
     */
    DoughnutGraph.prototype.graphElement;
    /**
     * @type {?}
     * @protected
     */
    DoughnutGraph.prototype.minSize;
    /**
     * @type {?}
     * @protected
     */
    DoughnutGraph.prototype.segmentsDegree;
}
