// @ngx-chart/doughnut/esm5/lib/doughnut-graph.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/doughnut-graph.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef } from '@angular/core';
import * as d3 from 'd3';
import { WrapperComponentD3, CircularMeasures } from '@ngx-chart/common';
import { DoughnutGraph } from './doughnut.class';
var DoughnutGraphComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DoughnutGraphComponent, _super);
    function DoughnutGraphComponent(elementRef, circularMeasures) {
        var _this = _super.call(this) || this;
        _this.elementRef = elementRef;
        _this.circularMeasures = circularMeasures;
        return _this;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    DoughnutGraphComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        // Remove previous graphs
        if (changes.graphData.previousValue) {
            d3.select("#svg-" + this.currentElementId).remove();
        }
        this.graphElement = new WrapperComponentD3(this.elementRef);
        this.currentElementId = this.graphElement.uid;
        this.minSize = this.circularMeasures.calcRadius(this.graphElement.width, this.graphElement.height);
        // If graph has labels and the position of if are extenal reduce radius in 25 points.
        this.minSize = this.graphData.labels && this.graphData.labels.position === 'external'
            ? this.minSize - 25
            : this.minSize;
        this.segmentsDegree = this.scaleData(this.graphData.values).map((/**
         * @param {?} value
         * @return {?}
         */
        function (value) { return _this.circularMeasures.arcValue(value); }));
        this.graphData.colors = this.checkGraphDataColors();
        this.segmentsDegree.map((/**
         * @param {?} arc
         * @param {?} idx
         * @return {?}
         */
        function (arc, idx) {
            /** @type {?} */
            var startAngle = idx === 0
                ? 0
                : _this.segmentsDegree.slice(0, idx).reduce((/**
                 * @param {?} curr
                 * @param {?} prev
                 * @return {?}
                 */
                function (curr, prev) { return curr + prev; }));
            /** @type {?} */
            var endAngle = idx === _this.segmentsDegree.length
                ? 360
                : _this.segmentsDegree.slice(0, idx + 1).reduce((/**
                 * @param {?} curr
                 * @param {?} prev
                 * @return {?}
                 */
                function (curr, prev) { return curr + prev; }));
            _this.drawSegment(startAngle, endAngle, idx);
        }));
    };
    DoughnutGraphComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-doughnut',
                    template: "",
                    providers: [
                        CircularMeasures
                    ]
                }] }
    ];
    /** @nocollapse */
    DoughnutGraphComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: CircularMeasures }
    ]; };
    return DoughnutGraphComponent;
}(DoughnutGraph));
export { DoughnutGraphComponent };
if (false) {
    /** @type {?} */
    DoughnutGraphComponent.prototype.currentElementId;
    /** @type {?} */
    DoughnutGraphComponent.prototype.elementRef;
    /** @type {?} */
    DoughnutGraphComponent.prototype.circularMeasures;
}
