// @ngx-chart/common/fesm5/ngx-chart-common.js
import { select, scaleLinear } from 'd3';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/wrapper-component-d3.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var WrapperComponentD3 = /** @class */ (function () {
    function WrapperComponentD3(_elementRef) {
        // Initialize private variables
        this._uid = this.generateUID();
        this._height = _elementRef.nativeElement.clientHeight;
        this._width = _elementRef.nativeElement.clientWidth;
        // Add Id to Element;
        _elementRef.nativeElement.id = this._uid;
        // Create svg container
        this._svgContainer = select("#" + this._uid)
            .append('svg')
            .attr('id', "svg-" + this._uid)
            .attr('height', this._height)
            .attr('width', this._width);
    }
    Object.defineProperty(WrapperComponentD3.prototype, "svgContainer", {
        /**
         * svgContainer
         * return svg container
         * @returns HTMLElement
         */
        get: /**
         * svgContainer
         * return svg container
         * @return {?} HTMLElement
         */
        function () {
            return this._svgContainer;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WrapperComponentD3.prototype, "uid", {
        /**
         * uid
         * return uid of container
         * @returns string
         */
        get: /**
         * uid
         * return uid of container
         * @return {?} string
         */
        function () {
            return this._uid;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WrapperComponentD3.prototype, "height", {
        /**
         * height
         * return height of container
         * @returns number
         */
        get: /**
         * height
         * return height of container
         * @return {?} number
         */
        function () {
            return this._height;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WrapperComponentD3.prototype, "width", {
        /**
         * width
         * return width of container
         * @returns number
         */
        get: /**
         * width
         * return width of container
         * @return {?} number
         */
        function () {
            return this._width;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * generateUID
     * Generate new UID. The uid is make with Date.now and plus random number and then replace it to string.
     * @returns string
     */
    /**
     * generateUID
     * Generate new UID. The uid is make with Date.now and plus random number and then replace it to string.
     * @private
     * @return {?} string
     */
    WrapperComponentD3.prototype.generateUID = /**
     * generateUID
     * Generate new UID. The uid is make with Date.now and plus random number and then replace it to string.
     * @private
     * @return {?} string
     */
    function () {
        return (Date.now() + Math.random()).toString(36).replace('.', '-');
    };
    return WrapperComponentD3;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/colors.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var Colors = /** @class */ (function () {
    function Colors() {
    }
    /**
     * generateRandomPallete
     *
     * Generate random palette of colors
     * @param length. Number of colors in generate pallete
     * @returns Array
     */
    /**
     * generateRandomPallete
     *
     * Generate random palette of colors
     * @param {?} length
     * @return {?} Array
     */
    Colors.prototype.generateRandomPallete = /**
     * generateRandomPallete
     *
     * Generate random palette of colors
     * @param {?} length
     * @return {?} Array
     */
    function (length) {
        /** @type {?} */
        var letters = '0123456789ABCDEF';
        /** @type {?} */
        var pallete = new Array(length).fill('');
        return pallete.map((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var color = '#';
            for (var i = 0; i < 6; i++) {
                color += letters[Math.floor(Math.random() * 16)];
            }
            return color;
        }));
    };
    return Colors;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: lib/circular-measures.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var CircularMeasures = /** @class */ (function () {
    function CircularMeasures() {
    }
    /**
     * calcRadius
     *
     * Calculate the minimum size between the height and width of the container element
     * with this value i will calculate the radius of circle.
     *
     * @param width width of container
     * @param height width of container
     * @returns number
     */
    /**
     * calcRadius
     *
     * Calculate the minimum size between the height and width of the container element
     * with this value i will calculate the radius of circle.
     *
     * @param {?} width width of container
     * @param {?} height width of container
     * @return {?} number
     */
    CircularMeasures.prototype.calcRadius = /**
     * calcRadius
     *
     * Calculate the minimum size between the height and width of the container element
     * with this value i will calculate the radius of circle.
     *
     * @param {?} width width of container
     * @param {?} height width of container
     * @return {?} number
     */
    function (width, height) {
        return width > height
            ? height
            : width;
    };
    /**
     * arcValue
     *
     * Calculate the arc to show.
     * Convert the pass value into a number that represent the degrees in a circle
     *
     * @param value value to represent
     * @returns number
     */
    /**
     * arcValue
     *
     * Calculate the arc to show.
     * Convert the pass value into a number that represent the degrees in a circle
     *
     * @param {?} value value to represent
     * @return {?} number
     */
    CircularMeasures.prototype.arcValue = /**
     * arcValue
     *
     * Calculate the arc to show.
     * Convert the pass value into a number that represent the degrees in a circle
     *
     * @param {?} value value to represent
     * @return {?} number
     */
    function (value) {
        /** @type {?} */
        var interValue = scaleLinear()
            .domain([0, 100])
            .range([0, 360]);
        return interValue(value);
    };
    /**
     * toRadians
     *
     * Convert degrees values in radians
     *
     * @param degree value of degree
     * @returns number
     */
    /**
     * toRadians
     *
     * Convert degrees values in radians
     *
     * @param {?} degree value of degree
     * @return {?} number
     */
    CircularMeasures.prototype.toRadians = /**
     * toRadians
     *
     * Convert degrees values in radians
     *
     * @param {?} degree value of degree
     * @return {?} number
     */
    function (degree) {
        return degree * Math.PI / 180;
    };
    return CircularMeasures;
}());

/**
 * @fileoverview added by tsickle
 * Generated from: public_api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: ngx-chart-common.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { WrapperComponentD3, Colors, CircularMeasures };
