// ng-primitives/esm2022/accordion/config/accordion.config.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { InjectionToken, inject } from '@angular/core';
export const defaultAccordionConfig = {
    type: 'single',
    collapsible: false,
    orientation: 'vertical',
};
export const NgpAccordionConfigToken = new InjectionToken('NgpAccordionConfigToken');
/**
 * Provide the default Accordion configuration
 * @param config The Accordion configuration
 * @returns The provider
 */
export function provideAccordionConfig(config) {
    return [
        {
            provide: NgpAccordionConfigToken,
            useValue: { ...defaultAccordionConfig, ...config },
        },
    ];
}
/**
 * Inject the Accordion configuration
 * @returns The global Accordion configuration
 */
export function injectAccordionConfig() {
    return inject(NgpAccordionConfigToken, { optional: true }) ?? defaultAccordionConfig;
}
