// ng-primitives/esm2022/radio/radio-indicator/radio-indicator.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, computed } from '@angular/core';
import { NgpHover, NgpPress } from 'ng-primitives/interactions';
import { injectRadioGroup } from '../radio-group/radio-group.token';
import { injectRadioItem } from '../radio-item/radio-item.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/interactions";
export class NgpRadioIndicator {
    constructor() {
        /**
         * Access the radio group.
         */
        this.radioGroup = injectRadioGroup();
        /**
         * Access the radio group item.
         */
        this.radioItem = injectRadioItem();
        /**
         * Determine if the radio indicator is checked.
         */
        this.checked = computed(() => this.radioGroup.value() === this.radioItem.value());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpRadioIndicator, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpRadioIndicator, isStandalone: true, selector: "[ngpRadioIndicator]", host: { properties: { "attr.data-checked": "checked() ? \"\" : null", "attr.data-disabled": "radioItem.disabled() ? \"\" : null" } }, hostDirectives: [{ directive: i1.NgpHover }, { directive: i1.NgpPress }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpRadioIndicator, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpRadioIndicator]',
                    standalone: true,
                    host: {
                        '[attr.data-checked]': 'checked() ? "" : null',
                        '[attr.data-disabled]': 'radioItem.disabled() ? "" : null',
                    },
                    hostDirectives: [NgpHover, NgpPress],
                }]
        }] });
