// ng-primitives/esm2022/pagination/pagination-button/pagination-button.directive.mjs
import { booleanAttribute, computed, Directive, HostListener, input, numberAttribute, } from '@angular/core';
import { NgpButton } from 'ng-primitives/button';
import { NgpDisabledToken } from 'ng-primitives/internal';
import { injectPagination } from '../pagination/pagination.token';
import { NgpPaginationButtonToken } from './pagination-button.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/button";
export class NgpPaginationButton {
    constructor() {
        /**
         * Access the pagination directive.
         */
        this.pagination = injectPagination();
        /**
         * Define the page this button represents.
         */
        this.page = input.required({
            alias: 'ngpPaginationButtonPage',
            transform: numberAttribute,
        });
        /**
         * Whether the button is disabled.
         */
        this.buttonDisabled = input(false, {
            alias: 'ngpPaginationButtonDisabled',
            transform: booleanAttribute,
        });
        /**
         * Whether the button is disabled.
         */
        this.disabled = computed(() => this.buttonDisabled() || this.pagination.disabled());
        /**
         * Whether this page is the currently selected page.
         */
        this.selected = computed(() => this.page() === this.pagination.page());
    }
    /**
     * Go to the page this button represents.
     */
    goToPage() {
        if (this.disabled()) {
            return;
        }
        this.pagination.page.set(this.page());
    }
    /**
     * A click event may not be fired if this is on an anchor tag and the href is empty.
     * This is a workaround to ensure the click event is fired.
     */
    onEnter(event) {
        event.preventDefault();
        event.stopPropagation();
        this.goToPage();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpPaginationButton, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpPaginationButton, isStandalone: true, selector: "[ngpPaginationButton]", inputs: { page: { classPropertyName: "page", publicName: "ngpPaginationButtonPage", isSignal: true, isRequired: true, transformFunction: null }, buttonDisabled: { classPropertyName: "buttonDisabled", publicName: "ngpPaginationButtonDisabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { listeners: { "click": "goToPage()", "keydown.enter": "onEnter($event)", "keydown.space": "onEnter($event)" }, properties: { "tabindex": "disabled() ? -1 : 0", "attr.data-disabled": "disabled() || pagination.disabled() ? \"\" : null", "attr.data-page": "page()", "attr.data-selected": "selected() ? \"\" : null", "attr.aria-current": "selected()" } }, providers: [
            { provide: NgpPaginationButtonToken, useExisting: NgpPaginationButton },
            { provide: NgpDisabledToken, useExisting: NgpPaginationButton },
        ], exportAs: ["ngpPaginationButton"], hostDirectives: [{ directive: i1.NgpButton }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpPaginationButton, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpPaginationButton]',
                    exportAs: 'ngpPaginationButton',
                    providers: [
                        { provide: NgpPaginationButtonToken, useExisting: NgpPaginationButton },
                        { provide: NgpDisabledToken, useExisting: NgpPaginationButton },
                    ],
                    hostDirectives: [NgpButton],
                    host: {
                        '[tabindex]': 'disabled() ? -1 : 0',
                        '[attr.data-disabled]': 'disabled() || pagination.disabled() ? "" : null',
                        '[attr.data-page]': 'page()',
                        '[attr.data-selected]': 'selected() ? "" : null',
                        '[attr.aria-current]': 'selected()',
                    },
                }]
        }], propDecorators: { goToPage: [{
                type: HostListener,
                args: ['click']
            }], onEnter: [{
                type: HostListener,
                args: ['keydown.enter', ['$event']]
            }, {
                type: HostListener,
                args: ['keydown.space', ['$event']]
            }] } });
