// ng-primitives/esm2022/avatar/avatar/avatar.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, signal } from '@angular/core';
import { NgpFocusVisible, NgpHover, NgpPress } from 'ng-primitives/interactions';
import { NgpAvatarToken } from './avatar.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/interactions";
export class NgpAvatar {
    constructor() {
        /**
         * Store the current state of the avatar.
         * @internal
         */
        this.state = signal(NgpAvatarState.Idle);
    }
    /**
     * Set the avatar state.
     * @param state The state to set.
     * @internal
     */
    setState(state) {
        this.state.set(state);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAvatar, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpAvatar, isStandalone: true, selector: "[ngpAvatar]", host: { properties: { "attr.data-state": "state()" } }, providers: [{ provide: NgpAvatarToken, useExisting: NgpAvatar }], hostDirectives: [{ directive: i1.NgpHover }, { directive: i1.NgpFocusVisible }, { directive: i1.NgpPress }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAvatar, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpAvatar]',
                    standalone: true,
                    providers: [{ provide: NgpAvatarToken, useExisting: NgpAvatar }],
                    host: {
                        '[attr.data-state]': 'state()',
                    },
                    hostDirectives: [NgpHover, NgpFocusVisible, NgpPress],
                }]
        }] });
export var NgpAvatarState;
(function (NgpAvatarState) {
    NgpAvatarState["Idle"] = "idle";
    NgpAvatarState["Loading"] = "loading";
    NgpAvatarState["Loaded"] = "loaded";
    NgpAvatarState["Error"] = "error";
})(NgpAvatarState || (NgpAvatarState = {}));
