// ng-primitives/esm2022/textarea/textarea/textarea.directive.mjs
import { booleanAttribute, Directive, input } from '@angular/core';
import { NgpFormControl } from 'ng-primitives/form-field';
import { NgpFocus, NgpHover, NgpPress } from 'ng-primitives/interactions';
import { NgpDisabledToken } from 'ng-primitives/internal';
import { NgpTextareaToken } from './textarea.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/form-field";
import * as i2 from "ng-primitives/interactions";
export class NgpTextarea {
    constructor() {
        /**
         * Whether the element is disabled.
         */
        this.disabled = input(false, {
            transform: booleanAttribute,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTextarea, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpTextarea, isStandalone: true, selector: "[ngpTextarea]", inputs: { disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null } }, providers: [
            { provide: NgpTextareaToken, useExisting: NgpTextarea },
            { provide: NgpDisabledToken, useExisting: NgpTextarea },
        ], exportAs: ["ngpTextarea"], hostDirectives: [{ directive: i1.NgpFormControl }, { directive: i2.NgpHover }, { directive: i2.NgpFocus }, { directive: i2.NgpPress }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTextarea, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpTextarea]',
                    exportAs: 'ngpTextarea',
                    providers: [
                        { provide: NgpTextareaToken, useExisting: NgpTextarea },
                        { provide: NgpDisabledToken, useExisting: NgpTextarea },
                    ],
                    hostDirectives: [NgpFormControl, NgpHover, NgpFocus, NgpPress],
                }]
        }] });
