// ng-primitives/esm2022/date-picker/date-picker-cell-render/date-picker-cell-render.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, inject, Injector, TemplateRef, ViewContainerRef, } from '@angular/core';
import { injectDatePickerWeek } from '../date-picker-row-render/date-picker-row-render.token';
import { NgpDatePickerCellDateToken, NgpDatePickerCellRenderToken, } from './date-picker-cell-render.token';
import * as i0 from "@angular/core";
export class NgpDatePickerCellRender {
    // Make sure the template checker knows the type of the context with which the
    // template of this directive will be rendered
    static ngTemplateContextGuard(_, context) {
        return true;
    }
    constructor() {
        /**
         * Access the template ref for the cell.
         */
        this.templateRef = inject(TemplateRef);
        /**
         * Access the view container ref.
         */
        this.viewContainerRef = inject(ViewContainerRef);
        /**
         * Access the dates in the week.
         */
        this.dates = injectDatePickerWeek();
        /**
         * Store the view refs for the dates.
         */
        this.viewRefs = [];
        this.renderDates();
    }
    /**
     * Render the dates in the week.
     */
    renderDates() {
        this.viewRefs.forEach(viewRef => viewRef.destroy());
        for (const date of this.dates) {
            const viewRef = this.viewContainerRef.createEmbeddedView(this.templateRef, {
                $implicit: date,
            }, {
                injector: Injector.create({
                    parent: this.viewContainerRef.injector,
                    providers: [{ provide: NgpDatePickerCellDateToken, useValue: date }],
                }),
            });
            this.viewRefs.push(viewRef);
        }
    }
    /**
     * Destroy the view refs.
     */
    ngOnDestroy() {
        this.viewRefs.forEach(viewRef => viewRef.destroy());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDatePickerCellRender, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpDatePickerCellRender, isStandalone: true, selector: "[ngpDatePickerCellRender]", providers: [{ provide: NgpDatePickerCellRenderToken, useExisting: NgpDatePickerCellRender }], exportAs: ["ngpDatePickerCellRender"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDatePickerCellRender, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpDatePickerCellRender]',
                    exportAs: 'ngpDatePickerCellRender',
                    providers: [{ provide: NgpDatePickerCellRenderToken, useExisting: NgpDatePickerCellRender }],
                }]
        }], ctorParameters: () => [] });
