// ng-primitives/esm2022/dialog/dialog-title/dialog-title.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, input } from '@angular/core';
import { onChange, uniqueId } from 'ng-primitives/utils';
import { injectDialog } from '../dialog/dialog.token';
import { NgpDialogTitleToken } from './dialog-title.token';
import * as i0 from "@angular/core";
export class NgpDialogTitle {
    constructor() {
        /** Access the dialog. */
        this.dialog = injectDialog();
        /** The id of the title. */
        this.id = input(uniqueId('ngp-dialog-title'));
        onChange(this.id, (id, prevId) => {
            if (prevId) {
                this.dialog.removeLabelledBy(prevId);
            }
            if (id) {
                this.dialog.setLabelledBy(id);
            }
        });
    }
    ngOnDestroy() {
        this.dialog.removeLabelledBy(this.id());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDialogTitle, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpDialogTitle, isStandalone: true, selector: "[ngpDialogTitle]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "id": "id()" } }, providers: [{ provide: NgpDialogTitleToken, useExisting: NgpDialogTitle }], exportAs: ["ngpDialogTitle"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDialogTitle, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpDialogTitle]',
                    exportAs: 'ngpDialogTitle',
                    providers: [{ provide: NgpDialogTitleToken, useExisting: NgpDialogTitle }],
                    host: {
                        '[id]': 'id()',
                    },
                }]
        }], ctorParameters: () => [] });
