// ng-primitives/esm2022/dialog/dialog/dialog-ref.mjs
import { hasModifierKey } from '@angular/cdk/keycodes';
import { inject } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * Reference to a dialog opened via the Dialog service.
 */
export class NgpDialogRef {
    constructor(overlayRef, config) {
        this.overlayRef = overlayRef;
        this.config = config;
        /** Emits when the dialog has been closed. */
        this.closed = new Subject();
        this.data = config.data;
        this.keydownEvents = overlayRef.keydownEvents();
        this.outsidePointerEvents = overlayRef.outsidePointerEvents();
        this.id = config.id; // By the time the dialog is created we are guaranteed to have an ID.
        this.keydownEvents.subscribe(event => {
            if (event.key === 'Escape' && !this.disableClose && !hasModifierKey(event)) {
                event.preventDefault();
                this.close('keyboard');
            }
        });
        this.detachSubscription = overlayRef.detachments().subscribe(() => this.close());
    }
    /**
     * Close the dialog.
     * @param result Optional result to return to the dialog opener.
     * @param options Additional options to customize the closing behavior.
     */
    close(focusOrigin) {
        this.overlayRef.dispose();
        this.detachSubscription.unsubscribe();
        this.closed.next(focusOrigin ?? null);
        this.closed.complete();
    }
    /** Updates the position of the dialog based on the current position strategy. */
    updatePosition() {
        this.overlayRef.updatePosition();
        return this;
    }
}
export function injectDialogRef() {
    return inject(NgpDialogRef);
}
