// @ng-matero/extensions/esm2022/grid/selectable-cell.mjs
import { Directive, EventEmitter, HostBinding, HostListener, Input, Output } from '@angular/core';
import * as i0 from "@angular/core";
export class MtxGridSelectableCell {
    constructor() {
        this.ctrlKeyPressed = false;
        this.shiftKeyPressed = false;
        this._selected = false;
        this.cellSelectable = true;
        this.cellSelectedChange = new EventEmitter();
    }
    get selected() {
        return this._selected;
    }
    onClick(event) {
        this.ctrlKeyPressed = event.ctrlKey;
        this.shiftKeyPressed = event.shiftKey;
        if (this.cellSelectable) {
            this.select();
        }
    }
    select() {
        this._selected = true;
        this.cellSelectedChange.emit(this);
    }
    deselect() {
        this._selected = false;
        this.cellSelectedChange.emit(this);
    }
    toggle() {
        this._selected = !this._selected;
        this.cellSelectedChange.emit(this);
    }
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxGridSelectableCell, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    /** @nocollapse */ static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.0", type: MtxGridSelectableCell, isStandalone: true, selector: "[mtx-grid-selectable-cell]", inputs: { cellSelectable: "cellSelectable" }, outputs: { cellSelectedChange: "cellSelectedChange" }, host: { listeners: { "click": "onClick($event)" }, properties: { "class.selected": "this.selected" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxGridSelectableCell, decorators: [{
            type: Directive,
            args: [{
                    selector: '[mtx-grid-selectable-cell]',
                    standalone: true,
                }]
        }], propDecorators: { selected: [{
                type: HostBinding,
                args: ['class.selected']
            }], cellSelectable: [{
                type: Input
            }], cellSelectedChange: [{
                type: Output
            }], onClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
