// ngx-progressbar/esm2022/http/ng-progress-http.directive.mjs
import { Directive, effect, inject, untracked } from '@angular/core';
import { NgProgressRef } from 'ngx-progressbar';
import { NgProgressHttpManager } from './ng-progress-http-manager';
import * as i0 from "@angular/core";
import * as i1 from "ngx-progressbar";
class NgProgressHttpBase {
    constructor() {
        this.manager = inject(NgProgressHttpManager);
        this.progressRef = inject(NgProgressRef, { host: true, self: true });
        let initial = true;
        effect(() => {
            const requestLoading = this.manager.requestsLoading();
            // Ignore the initial execution if loading state is false
            if (initial) {
                initial = false;
                if (!requestLoading)
                    return;
            }
            untracked(() => {
                if (requestLoading) {
                    this.progressRef.start();
                }
                else if (this.progressRef.active()) {
                    this.progressRef.complete();
                }
            });
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.12", ngImport: i0, type: NgProgressHttpBase, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.12", type: NgProgressHttpBase, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.12", ngImport: i0, type: NgProgressHttpBase, decorators: [{
            type: Directive
        }], ctorParameters: () => [] });
export class NgProgressHttp extends NgProgressHttpBase {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.12", ngImport: i0, type: NgProgressHttp, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.12", type: NgProgressHttp, isStandalone: true, selector: "ng-progress[ngProgressHttp]", usesInheritance: true, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.12", ngImport: i0, type: NgProgressHttp, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'ng-progress[ngProgressHttp]'
                }]
        }] });
export class NgProgressHttpRef extends NgProgressHttpBase {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.12", ngImport: i0, type: NgProgressHttpRef, deps: null, target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.12", type: NgProgressHttpRef, isStandalone: true, selector: "[ngProgressHttp]:not(ng-progress)", usesInheritance: true, hostDirectives: [{ directive: i1.NgProgressRef }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.12", ngImport: i0, type: NgProgressHttpRef, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngProgressHttp]:not(ng-progress)',
                    hostDirectives: [NgProgressRef]
                }]
        }] });
