// ngx-progressbar/esm2022/http/ng-progress-http-manager.mjs
import { Injectable, computed, inject } from '@angular/core';
import { NgProgressHttpCounter } from './ng-progress-http-counter';
import * as i0 from "@angular/core";
export class NgProgressHttpManager {
    constructor() {
        this.inProgressCount = inject(NgProgressHttpCounter);
        this.requestsCount = computed(() => this.inProgressCount());
        this.requestsLoading = computed(() => !!this.inProgressCount());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.12", ngImport: i0, type: NgProgressHttpManager, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.2.12", ngImport: i0, type: NgProgressHttpManager, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.12", ngImport: i0, type: NgProgressHttpManager, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }] });
