// ng-primitives/esm2022/slider/slider/slider.directive.mjs
import { Directive, booleanAttribute, computed, contentChild, input, model, numberAttribute, } from '@angular/core';
import { NgpFormControl } from 'ng-primitives/form-field';
import { NgpDisabledToken } from 'ng-primitives/internal';
import { NgpSliderTrackToken } from '../slider-track/slider-track.token';
import { NgpSliderToken } from './slider.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/form-field";
export class NgpSlider {
    constructor() {
        /**
         * The value of the slider.
         */
        this.value = model(0, {
            alias: 'ngpSliderValue',
        });
        /**
         * The minimum value of the slider.
         */
        this.min = input(0, {
            alias: 'ngpSliderMin',
            transform: numberAttribute,
        });
        /**
         * The maximum value of the slider.
         */
        this.max = input(100, {
            alias: 'ngpSliderMax',
            transform: numberAttribute,
        });
        /**
         * The step value of the slider.
         */
        this.step = input(1, {
            alias: 'ngpSliderStep',
            transform: numberAttribute,
        });
        /**
         * The orientation of the slider.
         */
        this.orientation = input('horizontal', {
            alias: 'ngpSliderOrientation',
        });
        /**
         * The disabled state of the slider.
         */
        this.disabled = input(false, {
            alias: 'ngpSliderDisabled',
            transform: booleanAttribute,
        });
        /**
         * Access the slider track.
         */
        this.track = contentChild(NgpSliderTrackToken);
        /**
         * The value as a percentage based on the min and max values.
         */
        this.percentage = computed(() => ((this.value() - this.min()) / (this.max() - this.min())) * 100);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSlider, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.2.0", version: "18.2.13", type: NgpSlider, isStandalone: true, selector: "[ngpSlider]", inputs: { value: { classPropertyName: "value", publicName: "ngpSliderValue", isSignal: true, isRequired: false, transformFunction: null }, min: { classPropertyName: "min", publicName: "ngpSliderMin", isSignal: true, isRequired: false, transformFunction: null }, max: { classPropertyName: "max", publicName: "ngpSliderMax", isSignal: true, isRequired: false, transformFunction: null }, step: { classPropertyName: "step", publicName: "ngpSliderStep", isSignal: true, isRequired: false, transformFunction: null }, orientation: { classPropertyName: "orientation", publicName: "ngpSliderOrientation", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpSliderDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { value: "ngpSliderValueChange" }, host: { properties: { "attr.data-orientation": "orientation()" } }, providers: [
            { provide: NgpSliderToken, useExisting: NgpSlider },
            { provide: NgpDisabledToken, useExisting: NgpSlider },
        ], queries: [{ propertyName: "track", first: true, predicate: NgpSliderTrackToken, descendants: true, isSignal: true }], exportAs: ["ngpSlider"], hostDirectives: [{ directive: i1.NgpFormControl }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSlider, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpSlider]',
                    exportAs: 'ngpSlider',
                    providers: [
                        { provide: NgpSliderToken, useExisting: NgpSlider },
                        { provide: NgpDisabledToken, useExisting: NgpSlider },
                    ],
                    hostDirectives: [NgpFormControl],
                    host: {
                        '[attr.data-orientation]': 'orientation()',
                    },
                }]
        }] });
