// ng-primitives/esm2022/radio/radio-group/radio-group.directive.mjs
import { Directive, booleanAttribute, input, model } from '@angular/core';
import { NgpFormControl } from 'ng-primitives/form-field';
import { NgpDisabledToken } from 'ng-primitives/internal';
import { NgpRovingFocusGroup } from 'ng-primitives/roving-focus';
import { NgpRadioGroupToken } from './radio-group.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/roving-focus";
import * as i2 from "ng-primitives/form-field";
export class NgpRadioGroup {
    constructor() {
        /**
         * The value of the radio group.
         */
        this.value = model(null, { alias: 'ngpRadioGroupValue' });
        /**
         * Whether the radio group is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpRadioGroupDisabled',
            transform: booleanAttribute,
        });
        /**
         * The orientation of the radio group.
         * @default 'horizontal'
         */
        this.orientation = input('horizontal', {
            alias: 'ngpRadioGroupOrientation',
        });
    }
    /**
     * Select a radio item.
     * @param value The value of the radio item to select.
     */
    select(value) {
        this.value.set(value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpRadioGroup, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpRadioGroup, isStandalone: true, selector: "[ngpRadioGroup]", inputs: { value: { classPropertyName: "value", publicName: "ngpRadioGroupValue", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpRadioGroupDisabled", isSignal: true, isRequired: false, transformFunction: null }, orientation: { classPropertyName: "orientation", publicName: "ngpRadioGroupOrientation", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { value: "ngpRadioGroupValueChange" }, host: { attributes: { "role": "radiogroup" }, properties: { "attr.aria-orientation": "orientation()", "attr.data-orientation": "orientation()" } }, providers: [
            { provide: NgpRadioGroupToken, useExisting: NgpRadioGroup },
            { provide: NgpDisabledToken, useExisting: NgpRadioGroup },
        ], hostDirectives: [{ directive: i1.NgpRovingFocusGroup }, { directive: i2.NgpFormControl }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpRadioGroup, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpRadioGroup]',
                    standalone: true,
                    providers: [
                        { provide: NgpRadioGroupToken, useExisting: NgpRadioGroup },
                        { provide: NgpDisabledToken, useExisting: NgpRadioGroup },
                    ],
                    hostDirectives: [NgpRovingFocusGroup, NgpFormControl],
                    host: {
                        role: 'radiogroup',
                        '[attr.aria-orientation]': 'orientation()',
                        '[attr.data-orientation]': 'orientation()',
                    },
                }]
        }] });
