// ng-primitives/esm2022/search/search-field-clear/search-field-clear.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, HostListener } from '@angular/core';
import { injectSearchField } from '../search-field/search-field.token';
import { NgpSearchFieldClearToken } from './search-field-clear.token';
import * as i0 from "@angular/core";
export class NgpSearchFieldClear {
    constructor() {
        /**
         * Access the Search Field instance.
         */
        this.searchField = injectSearchField();
    }
    /**
     * Clear the input field.
     */
    click() {
        this.searchField.clear();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSearchFieldClear, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpSearchFieldClear, isStandalone: true, selector: "[ngpSearchFieldClear]", host: { listeners: { "click": "click()" }, properties: { "tabindex": "-1", "attr.data-empty": "searchField.empty() ? \"\" : null" } }, providers: [{ provide: NgpSearchFieldClearToken, useExisting: NgpSearchFieldClear }], exportAs: ["ngpSearchFieldClear"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSearchFieldClear, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpSearchFieldClear]',
                    exportAs: 'ngpSearchFieldClear',
                    providers: [{ provide: NgpSearchFieldClearToken, useExisting: NgpSearchFieldClear }],
                    host: {
                        '[tabindex]': '-1',
                        '[attr.data-empty]': 'searchField.empty() ? "" : null',
                    },
                }]
        }], propDecorators: { click: [{
                type: HostListener,
                args: ['click']
            }] } });
