// ng-primitives/esm2022/avatar/avatar-fallback/avatar-fallback.directive.mjs
import { Directive, computed, input, numberAttribute, signal } from '@angular/core';
import { injectDisposables } from 'ng-primitives/utils';
import { NgpAvatarState } from '../avatar/avatar.directive';
import { injectAvatar } from '../avatar/avatar.token';
import { injectAvatarConfig } from '../config/avatar.config';
import * as i0 from "@angular/core";
export class NgpAvatarFallback {
    constructor() {
        /**
         * Access the avatar
         */
        this.avatar = injectAvatar();
        /**
         * Access the global configuration.
         */
        this.config = injectAvatarConfig();
        /**
         * Access the disposable utilities.
         */
        this.disposables = injectDisposables();
        /**
         * Define a delay before the fallback is shown. This is useful to only show the fallback for those with slower connections.
         * @default 0
         */
        this.delay = input(this.config.delay, {
            alias: 'ngpAvatarFallbackDelay',
            transform: numberAttribute,
        });
        /**
         * Determine if this element should be hidden.
         * @returns True if the element should be visible
         */
        this.visible = computed(() => 
        // we need to check if the element can render and if the avatar is not in a loaded state
        this.delayElapsed() && this.avatar.state() !== NgpAvatarState.Loaded);
        /**
         * Determine the delay has elapsed, and we can show the fallback.
         */
        this.delayElapsed = signal(false);
    }
    ngOnInit() {
        this.disposables.setTimeout(() => this.delayElapsed.set(true), this.delay());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAvatarFallback, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpAvatarFallback, isStandalone: true, selector: "[ngpAvatarFallback]", inputs: { delay: { classPropertyName: "delay", publicName: "ngpAvatarFallbackDelay", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "style.display": "visible() ? null : \"none\"" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAvatarFallback, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngpAvatarFallback]',
                    standalone: true,
                    host: {
                        '[style.display]': 'visible() ? null : "none"',
                    },
                }]
        }] });
