// ng-primitives/esm2022/dialog/dialog-trigger/dialog-trigger.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { FocusMonitor } from '@angular/cdk/a11y';
import { ApplicationRef, Directive, ElementRef, HostListener, inject, input, ViewContainerRef, } from '@angular/core';
import { NgpDialogManager } from '../dialog/dialog.service';
import { NgpDialogTriggerToken } from './dialog-trigger.token';
import * as i0 from "@angular/core";
export class NgpDialogTrigger {
    constructor() {
        /** Access the dialog manager. */
        this.dialogManager = inject(NgpDialogManager);
        /** Access the application ref. */
        this.applicationRef = inject(ApplicationRef);
        /** Access the focus monitor. */
        this.focusMonitor = inject(FocusMonitor);
        /** Access the element ref. */
        this.elementRef = inject(ElementRef);
        /** The template to launch. */
        this.template = input.required({
            alias: 'ngpDialogTrigger',
        });
        /**
         * Store the dialog ref.
         * @internal
         */
        this.dialogRef = null;
    }
    launch() {
        // this is not ideal, but there is a case where a dialog trigger is within an overlay (e.g. menu),
        // which may be removed before the dialog is closed. This is not desired, so we need to access a view container ref
        // that is not within the overlay. To solve this we use the view container ref of the root component.
        // Could this have any unintended side effects? For example, the dialog would not be closed during route changes?
        const viewContainerRef = this.applicationRef.components[0].injector.get(ViewContainerRef);
        this.dialogRef = this.dialogManager.open(this.template(), {
            viewContainerRef,
        });
        this.dialogRef.closed.subscribe(focusOrigin => {
            this.dialogRef = null;
            // Focus the trigger element after the dialog closes.
            this.focusMonitor.focusVia(this.elementRef.nativeElement, focusOrigin);
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDialogTrigger, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpDialogTrigger, isStandalone: true, selector: "[ngpDialogTrigger]", inputs: { template: { classPropertyName: "template", publicName: "ngpDialogTrigger", isSignal: true, isRequired: true, transformFunction: null } }, host: { listeners: { "click": "launch()" } }, providers: [{ provide: NgpDialogTriggerToken, useExisting: NgpDialogTrigger }], exportAs: ["ngpDialogTrigger"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDialogTrigger, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpDialogTrigger]',
                    exportAs: 'ngpDialogTrigger',
                    providers: [{ provide: NgpDialogTriggerToken, useExisting: NgpDialogTrigger }],
                }]
        }], propDecorators: { launch: [{
                type: HostListener,
                args: ['click']
            }] } });
