// ng-primitives/esm2022/switch/switch/switch.directive.mjs
import { Directive, ElementRef, HostListener, booleanAttribute, inject, input, model, } from '@angular/core';
import { NgpFormControl } from 'ng-primitives/form-field';
import { NgpFocusVisible, NgpHover, NgpPress } from 'ng-primitives/interactions';
import { NgpDisabledToken } from 'ng-primitives/internal';
import { NgpSwitchToken } from './switch.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/form-field";
import * as i2 from "ng-primitives/interactions";
export class NgpSwitch {
    constructor() {
        /**
         * Access the element ref.
         */
        this.elementRef = inject(ElementRef);
        /**
         * Determine if the switch is a button
         */
        this.isButton = this.elementRef.nativeElement.tagName === 'BUTTON';
        /**
         * Determine if the switch is checked.
         * @default false
         */
        this.checked = model(false, {
            alias: 'ngpSwitchChecked',
        });
        /**
         * Determine if the switch is disabled.
         * @default false
         */
        this.disabled = input(false, {
            alias: 'ngpSwitchDisabled',
            transform: booleanAttribute,
        });
    }
    /**
     * Toggle the checked state.
     */
    toggle() {
        if (this.disabled()) {
            return;
        }
        this.checked.update(checked => !checked);
    }
    /**
     * Handle the keydown event.
     */
    onKeyDown() {
        // If the switch is not a button then the space key will not toggle the checked state automatically,
        // so we need to do it manually.
        if (!this.isButton) {
            this.toggle();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSwitch, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpSwitch, isStandalone: true, selector: "[ngpSwitch]", inputs: { checked: { classPropertyName: "checked", publicName: "ngpSwitchChecked", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpSwitchDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { checked: "ngpSwitchCheckedChange" }, host: { attributes: { "role": "switch" }, listeners: { "click": "toggle()", "keydown.space": "onKeyDown()" }, properties: { "attr.type": "isButton ? \"button\" : null", "attr.aria-checked": "checked()", "attr.data-checked": "checked() ? \"\" : null", "attr.disabled": "isButton && disabled() ? \"\" : null" } }, providers: [
            { provide: NgpSwitchToken, useExisting: NgpSwitch },
            { provide: NgpDisabledToken, useExisting: NgpSwitch },
        ], exportAs: ["ngpSwitch"], hostDirectives: [{ directive: i1.NgpFormControl }, { directive: i2.NgpHover }, { directive: i2.NgpPress }, { directive: i2.NgpFocusVisible }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSwitch, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpSwitch]',
                    exportAs: 'ngpSwitch',
                    providers: [
                        { provide: NgpSwitchToken, useExisting: NgpSwitch },
                        { provide: NgpDisabledToken, useExisting: NgpSwitch },
                    ],
                    hostDirectives: [NgpFormControl, NgpHover, NgpPress, NgpFocusVisible],
                    host: {
                        role: 'switch',
                        '[attr.type]': 'isButton ? "button" : null',
                        '[attr.aria-checked]': 'checked()',
                        '[attr.data-checked]': 'checked() ? "" : null',
                        '[attr.disabled]': 'isButton && disabled() ? "" : null',
                    },
                }]
        }], propDecorators: { toggle: [{
                type: HostListener,
                args: ['click']
            }], onKeyDown: [{
                type: HostListener,
                args: ['keydown.space']
            }] } });
