// ng-primitives/esm2022/interactions/focus/focus.directive.mjs
import { Directive, booleanAttribute, input, output } from '@angular/core';
import { injectDisabled, setupFocus } from 'ng-primitives/internal';
import { NgpFocusToken } from './focus.token';
import * as i0 from "@angular/core";
/**
 * This was inspired by the React Aria useFocus hook.
 * https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/interactions/src/useFocus.ts#L20
 */
export class NgpFocus {
    constructor() {
        /**
         * Whether listening for focus events is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpFocusDisabled',
            transform: booleanAttribute,
        });
        /**
         * Access the disabled state from any parent.
         */
        this.isDisabled = injectDisabled(this.disabled);
        /**
         * Emit when the focus state changes.
         */
        this.focus = output({ alias: 'ngpFocus' });
        // setup the focus listener
        setupFocus({
            disabled: this.isDisabled,
            focus: () => this.focus.emit(true),
            blur: () => this.focus.emit(false),
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpFocus, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpFocus, isStandalone: true, selector: "[ngpFocus]", inputs: { disabled: { classPropertyName: "disabled", publicName: "ngpFocusDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { focus: "ngpFocus" }, providers: [{ provide: NgpFocusToken, useExisting: NgpFocus }], exportAs: ["ngpFocus"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpFocus, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpFocus]',
                    exportAs: 'ngpFocus',
                    providers: [{ provide: NgpFocusToken, useExisting: NgpFocus }],
                }]
        }], ctorParameters: () => [] });
