// ng-primitives/esm2022/popover/popover/popover.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, ElementRef, computed, inject, isDevMode } from '@angular/core';
import { NgpFocusTrap } from 'ng-primitives/focus-trap';
import { injectPopoverTrigger } from '../popover-trigger/popover-trigger.token';
import { NgpPopoverToken } from './popover.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/focus-trap";
export class NgpPopover {
    constructor() {
        /**
         * Access the popover element.
         */
        this.popover = inject((ElementRef));
        /**
         * Access the trigger instance.
         */
        this.trigger = injectPopoverTrigger();
        /**
         * Compute the x position of the popover.
         */
        this.x = computed(() => this.trigger.position().x);
        /**
         * Compute the y position of the popover.
         */
        this.y = computed(() => this.trigger.position().y);
    }
    ngOnInit() {
        // if the element does not have a fixed position then throw a warning in dev mode
        if (isDevMode()) {
            const { position } = getComputedStyle(this.popover.nativeElement);
            if (position !== 'absolute') {
                console.warn(`The popover element must have an absolute position. The current position is ${position}.`);
            }
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpPopover, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpPopover, isStandalone: true, selector: "[ngpPopover]", host: { attributes: { "role": "menu" }, properties: { "style.left.px": "x()", "style.top.px": "y()" } }, providers: [{ provide: NgpPopoverToken, useExisting: NgpPopover }], exportAs: ["ngpPopover"], hostDirectives: [{ directive: i1.NgpFocusTrap }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpPopover, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpPopover]',
                    exportAs: 'ngpPopover',
                    hostDirectives: [NgpFocusTrap],
                    providers: [{ provide: NgpPopoverToken, useExisting: NgpPopover }],
                    host: {
                        role: 'menu',
                        '[style.left.px]': 'x()',
                        '[style.top.px]': 'y()',
                    },
                }]
        }] });
