// ng-primitives/esm2022/interactions/hover/hover.directive.mjs
import { Directive, booleanAttribute, input, output } from '@angular/core';
import { injectDisabled, setupHover } from 'ng-primitives/internal';
import { NgpHoverToken } from './hover.token';
import * as i0 from "@angular/core";
/**
 * A directive for normalizing hover events across the different browsers and devices.
 *
 * This is an Angular port of the useHover hook from
 * react-aria: https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/interactions/src/useHover.ts
 */
export class NgpHover {
    /**
     * Setup the hover state.
     */
    constructor() {
        /**
         * Whether hoving should be disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpHoverDisabled',
            transform: booleanAttribute,
        });
        /**
         * Access the disabled state from any parent.
         */
        this.isDisabled = injectDisabled(this.disabled);
        /**
         * Emit an event when hovering starts.
         */
        this.hoverStart = output({ alias: 'ngpHoverStart' });
        /**
         * Emit an event when hovering ends.
         */
        this.hoverEnd = output({ alias: 'ngpHoverEnd' });
        /**
         * Emit an event when the hover state changes.
         */
        this.hoverChange = output({ alias: 'ngpHover' });
        // setup the hover listener
        setupHover({
            hoverStart: () => {
                this.hoverStart.emit();
                this.hoverChange.emit(true);
            },
            hoverEnd: () => {
                this.hoverEnd.emit();
                this.hoverChange.emit(false);
            },
            disabled: this.isDisabled,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpHover, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpHover, isStandalone: true, selector: "[ngpHover]", inputs: { disabled: { classPropertyName: "disabled", publicName: "ngpHoverDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { hoverStart: "ngpHoverStart", hoverEnd: "ngpHoverEnd", hoverChange: "ngpHover" }, providers: [{ provide: NgpHoverToken, useExisting: NgpHover }], exportAs: ["ngpHover"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpHover, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpHover]',
                    exportAs: 'ngpHover',
                    providers: [{ provide: NgpHoverToken, useExisting: NgpHover }],
                }]
        }], ctorParameters: () => [] });
