// ng-primitives/esm2022/select/select/select.directive.mjs
import { booleanAttribute, Directive, input } from '@angular/core';
import { NgpFormControl } from 'ng-primitives/form-field';
import { NgpFocusVisible, NgpHover, NgpPress } from 'ng-primitives/interactions';
import { NgpDisabledToken } from 'ng-primitives/internal';
import { NgpSelectToken } from './select.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/form-field";
import * as i2 from "ng-primitives/interactions";
export class NgpSelect {
    constructor() {
        /**
         * Whether the select is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpSelectDisabled',
            transform: booleanAttribute,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSelect, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpSelect, isStandalone: true, selector: "select[ngpSelect]", inputs: { disabled: { classPropertyName: "disabled", publicName: "ngpSelectDisabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "attr.disabled": "disabled() || null" } }, providers: [
            { provide: NgpSelectToken, useExisting: NgpSelect },
            { provide: NgpDisabledToken, useExisting: NgpSelect },
        ], exportAs: ["ngpSelect"], hostDirectives: [{ directive: i1.NgpFormControl }, { directive: i2.NgpFocusVisible }, { directive: i2.NgpHover }, { directive: i2.NgpPress }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpSelect, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: 'select[ngpSelect]',
                    exportAs: 'ngpSelect',
                    providers: [
                        { provide: NgpSelectToken, useExisting: NgpSelect },
                        { provide: NgpDisabledToken, useExisting: NgpSelect },
                    ],
                    hostDirectives: [NgpFormControl, NgpFocusVisible, NgpHover, NgpPress],
                    host: {
                        '[attr.disabled]': 'disabled() || null',
                    },
                }]
        }] });
