// ng-primitives/esm2022/accordion/accordion-item/accordion-item.directive.mjs
import { Directive, booleanAttribute, computed, contentChild, input } from '@angular/core';
import { NgpAccordionContentToken } from '../accordion-content/accordion-content.token';
import { NgpAccordionTriggerToken } from '../accordion-trigger/accordion-trigger.token';
import { injectAccordion } from '../accordion/accordion.token';
import { NgpAccordionItemToken } from './accordion-item.token';
import * as i0 from "@angular/core";
export class NgpAccordionItem {
    constructor() {
        /**
         * Access the accordion.
         */
        this.accordion = injectAccordion();
        /**
         * The value of the accordion item.
         */
        this.value = input.required({
            alias: 'ngpAccordionItemValue',
        });
        /**
         * Whether the accordion item is disabled.
         */
        this.disabled = input(false, {
            alias: 'ngpAccordionItemDisabled',
            transform: booleanAttribute,
        });
        /**
         * Access the accordion trigger
         */
        this.trigger = contentChild(NgpAccordionTriggerToken);
        /**
         * Access the accordion content
         */
        this.content = contentChild(NgpAccordionContentToken);
        /**
         * Whether the accordion item is expanded.
         */
        this.open = computed(() => this.accordion.isOpen(this.value()));
        /**
         * The trigger id.
         */
        this.triggerId = computed(() => this.trigger()?.id());
        /**
         * The content id.
         */
        this.contentId = computed(() => this.content()?.id());
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAccordionItem, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.2.0", version: "18.2.13", type: NgpAccordionItem, isStandalone: true, selector: "[ngpAccordionItem]", inputs: { value: { classPropertyName: "value", publicName: "ngpAccordionItemValue", isSignal: true, isRequired: true, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpAccordionItemDisabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "attr.data-orientation": "accordion.orientation()", "attr.data-open": "open() ? \"\" : null", "attr.data-disabled": "disabled() || accordion.disabled() ? \"\" : null" } }, providers: [{ provide: NgpAccordionItemToken, useExisting: NgpAccordionItem }], queries: [{ propertyName: "trigger", first: true, predicate: NgpAccordionTriggerToken, descendants: true, isSignal: true }, { propertyName: "content", first: true, predicate: NgpAccordionContentToken, descendants: true, isSignal: true }], exportAs: ["ngpAccordionItem"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAccordionItem, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpAccordionItem]',
                    exportAs: 'ngpAccordionItem',
                    providers: [{ provide: NgpAccordionItemToken, useExisting: NgpAccordionItem }],
                    host: {
                        '[attr.data-orientation]': 'accordion.orientation()',
                        '[attr.data-open]': 'open() ? "" : null',
                        '[attr.data-disabled]': 'disabled() || accordion.disabled() ? "" : null',
                    },
                }]
        }] });
