// ng-primitives/esm2022/tabs/tab-list/tab-list.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive } from '@angular/core';
import { injectTabset } from '../tabset/tabset.token';
import * as i0 from "@angular/core";
export class NgpTabList {
    constructor() {
        /**
         * Access the tabset
         */
        this.tabset = injectTabset();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTabList, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpTabList, isStandalone: true, selector: "[ngpTabList]", host: { attributes: { "role": "tablist" }, properties: { "attr.aria-orientation": "tabset.orientation()", "attr.data-orientation": "tabset.orientation()" } }, exportAs: ["ngpTabList"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTabList, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpTabList]',
                    exportAs: 'ngpTabList',
                    host: {
                        role: 'tablist',
                        '[attr.aria-orientation]': 'tabset.orientation()',
                        '[attr.data-orientation]': 'tabset.orientation()',
                    },
                }]
        }] });
