// ng-primitives/esm2022/interactions/focus-visible/focus-visible.directive.mjs
import { Directive, booleanAttribute, input, output } from '@angular/core';
import { injectDisabled, setupFocusVisible } from 'ng-primitives/internal';
import { NgpFocusVisibleToken } from './focus-visible.token';
import * as i0 from "@angular/core";
export class NgpFocusVisible {
    constructor() {
        /**
         * Whether focus events are listened to.
         */
        this.disabled = input(false, {
            alias: 'ngpFocusVisibleDisabled',
            transform: booleanAttribute,
        });
        /**
         * Access the disabled state from any parent.
         */
        this.isDisabled = injectDisabled(this.disabled);
        /**
         * Emit when the element is visually focused.
         */
        this.focusChange = output({
            alias: 'ngpFocusVisible',
        });
        // setup the focus visible listener
        setupFocusVisible({
            disabled: this.isDisabled,
            focusChange: value => this.focusChange.emit(value),
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpFocusVisible, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpFocusVisible, isStandalone: true, selector: "[ngpFocusVisible]", inputs: { disabled: { classPropertyName: "disabled", publicName: "ngpFocusVisibleDisabled", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { focusChange: "ngpFocusVisible" }, providers: [{ provide: NgpFocusVisibleToken, useExisting: NgpFocusVisible }], exportAs: ["ngpFocusVisible"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpFocusVisible, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpFocusVisible]',
                    exportAs: 'ngpFocusVisible',
                    providers: [{ provide: NgpFocusVisibleToken, useExisting: NgpFocusVisible }],
                }]
        }], ctorParameters: () => [] });
