// ng-primitives/esm2022/tabs/tab-button/tab-button.directive.mjs
import { Directive, HostListener, booleanAttribute, computed, input } from '@angular/core';
import { NgpFocusVisible, NgpHover, NgpPress } from 'ng-primitives/interactions';
import { NgpRovingFocusItem } from 'ng-primitives/roving-focus';
import { injectTabset } from '../tabset/tabset.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/roving-focus";
import * as i2 from "ng-primitives/interactions";
export class NgpTabButton {
    constructor() {
        /**
         * Access the tabset
         */
        this.tabset = injectTabset();
        /**
         * The value of the tab this trigger controls
         */
        this.value = input.required({ alias: 'ngpTabButtonValue' });
        /**
         * Whether the tab is disabled
         * @default false
         */
        this.disabled = input(false, {
            alias: 'ngpTabButtonDisabled',
            transform: booleanAttribute,
        });
        /**
         * Determine the id of the tab button
         * @internal
         */
        this.id = input();
        /**
         * Determine a unique id for the tab button if not provided
         * @internal
         */
        this.defaultId = computed(() => `${this.tabset.id()}-button-${this.value()}`);
        /**
         * Determine the aria-controls of the tab button
         * @internal
         */
        this.ariaControls = computed(() => `${this.tabset.id()}-panel-${this.value()}`);
        /**
         * Whether the tab is active
         */
        this.active = computed(() => this.tabset.value() === this.value());
    }
    /**
     * Select the tab this trigger controls
     */
    select() {
        this.tabset.select(this.value());
    }
    /**
     * On focus select the tab this trigger controls if activateOnFocus is true
     */
    activateOnFocus() {
        if (this.tabset.activateOnFocus()) {
            this.select();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTabButton, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpTabButton, isStandalone: true, selector: "[ngpTabButton]", inputs: { value: { classPropertyName: "value", publicName: "ngpTabButtonValue", isSignal: true, isRequired: true, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "ngpTabButtonDisabled", isSignal: true, isRequired: false, transformFunction: null }, id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "tab" }, listeners: { "click": "select()", "focus": "activateOnFocus()" }, properties: { "attr.id": "id() ?? defaultId()", "attr.aria-controls": "ariaControls()", "attr.data-active": "active() ? \"\" : null", "attr.data-disabled": "disabled() ? \"\" : null", "attr.data-orientation": "tabset.orientation()" } }, exportAs: ["ngpTabButton"], hostDirectives: [{ directive: i1.NgpRovingFocusItem }, { directive: i2.NgpHover }, { directive: i2.NgpFocusVisible }, { directive: i2.NgpPress }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpTabButton, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpTabButton]',
                    exportAs: 'ngpTabButton',
                    host: {
                        role: 'tab',
                        '[attr.id]': 'id() ?? defaultId()',
                        '[attr.aria-controls]': 'ariaControls()',
                        '[attr.data-active]': 'active() ? "" : null',
                        '[attr.data-disabled]': 'disabled() ? "" : null',
                        '[attr.data-orientation]': 'tabset.orientation()',
                    },
                    hostDirectives: [NgpRovingFocusItem, NgpHover, NgpFocusVisible, NgpPress],
                }]
        }], propDecorators: { select: [{
                type: HostListener,
                args: ['click']
            }], activateOnFocus: [{
                type: HostListener,
                args: ['focus']
            }] } });
