// ng-primitives/esm2022/button/button/button.directive.mjs
import { booleanAttribute, Directive, ElementRef, inject, input } from '@angular/core';
import { NgpDisabledToken, setupButton } from 'ng-primitives/internal';
import { NgpButtonToken } from './button.token';
import * as i0 from "@angular/core";
export class NgpButton {
    constructor() {
        /**
         * Get the native element of the button.
         */
        this.elementRef = inject(ElementRef);
        /**
         * Whether the button is disabled.
         */
        this.disabled = input(false, {
            transform: booleanAttribute,
        });
        /**
         * Detect if this is an HTML button element.
         */
        this.isButton = this.elementRef.nativeElement.tagName.toLowerCase() === 'button';
        // setup the hover, press, and focus-visible listeners
        setupButton({ disabled: this.disabled });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpButton, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpButton, isStandalone: true, selector: "[ngpButton]", inputs: { disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "attr.data-disabled": "disabled() ? \"\" : null", "attr.disabled": "isButton && disabled() ? true : null" } }, providers: [
            { provide: NgpButtonToken, useExisting: NgpButton },
            { provide: NgpDisabledToken, useExisting: NgpButton },
        ], exportAs: ["ngpButton"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpButton, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpButton]',
                    exportAs: 'ngpButton',
                    providers: [
                        { provide: NgpButtonToken, useExisting: NgpButton },
                        { provide: NgpDisabledToken, useExisting: NgpButton },
                    ],
                    host: {
                        '[attr.data-disabled]': 'disabled() ? "" : null',
                        '[attr.disabled]': 'isButton && disabled() ? true : null',
                    },
                }]
        }], ctorParameters: () => [] });
