// ng-primitives/esm2022/toast/config/toast.config.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { InjectionToken, inject } from '@angular/core';
export const defaultToastConfig = {
    gap: 16,
    duration: 3000,
    gravity: 'top',
    position: 'end',
    stopOnHover: true,
    ariaLive: 'polite',
};
export const NgpToastConfigToken = new InjectionToken('NgpToastConfigToken');
/**
 * Provide the default Toast configuration
 * @param config The Toast configuration
 * @returns The provider
 */
export function provideToastConfig(config) {
    return [
        {
            provide: NgpToastConfigToken,
            useValue: { ...defaultToastConfig, ...config },
        },
    ];
}
/**
 * Inject the Toast configuration
 * @returns The global Toast configuration
 */
export function injectToastConfig() {
    return inject(NgpToastConfigToken, { optional: true }) ?? defaultToastConfig;
}
