// @ngx-formly/material/esm2020/slider/slider-mdc.type.mjs
import { Component, ChangeDetectionStrategy, ViewChild } from '@angular/core';
import { MatSlider } from '@angular/material/slider';
import { MAT_SLIDER_THUMB } from './slider-interface';
import { FormlyFieldSlider } from './slider.type';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/slider";
import * as i2 from "@ngx-formly/core";
import * as i3 from "./slider-input";
import * as i4 from "@angular/forms";
/**
 * @ignore
 */
export class FormlyFieldMDCSlider extends FormlyFieldSlider {
    set mdcSlider(slider) {
        Object.defineProperty(slider, '_input', {
            set: () => { },
            get: () => this.sliderThumb,
        });
        const visualThumb = {
            _hostElement: {
                classList: {
                    add: () => { },
                    remove: () => { },
                },
            },
        };
        // workarround for "ERROR TypeError: visualThumb is undefined" when `discrete` is set to `true`
        slider._thumbs = { first: visualThumb, last: visualThumb };
    }
    onContainerClick(event) {
        this.sliderThumb?._hostElement.focus();
        super.onContainerClick(event);
    }
}
FormlyFieldMDCSlider.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: FormlyFieldMDCSlider, deps: null, target: i0.ɵɵFactoryTarget.Component });
FormlyFieldMDCSlider.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.12", type: FormlyFieldMDCSlider, selector: "formly-field-mat-slider", viewQueries: [{ propertyName: "sliderThumb", first: true, predicate: MAT_SLIDER_THUMB, descendants: true, static: true }, { propertyName: "mdcSlider", first: true, predicate: MatSlider, descendants: true, static: true }], usesInheritance: true, ngImport: i0, template: `
    <mat-slider
      [id]="id"
      [style.width]="'100%'"
      [formlyAttributes]="field"
      [tabIndex]="props.tabindex"
      [color]="props.color"
      [displayWith]="props.displayWith"
      [invert]="props.invert"
      [max]="props.max"
      [min]="props.min"
      [step]="props.step"
      [discrete]="props.thumbLabel || props.discrete"
      [showTickMarks]="props.showTickMarks"
      [step]="props.step"
      [tickInterval]="props.tickInterval"
      [valueText]="props.valueText"
      [vertical]="props.vertical"
      (input)="props.input && props.input(field, $event)"
      (change)="props.change && props.change(field, $event)"
    >
      <input matSliderThumb [formControl]="formControl" [formlyAttributes]="field" />
    </mat-slider>
  `, isInline: true, components: [{ type: i1.MatSlider, selector: "mat-slider", inputs: ["disabled", "color", "tabIndex", "invert", "max", "min", "step", "thumbLabel", "tickInterval", "value", "displayWith", "valueText", "vertical"], outputs: ["change", "input", "valueChange"], exportAs: ["matSlider"] }], directives: [{ type: i2.ɵFormlyAttributes, selector: "[formlyAttributes]", inputs: ["formlyAttributes", "id"] }, { type: i3.MatSliderThumb, selector: "input[matSliderThumb]", inputs: ["value"], outputs: ["valueChange", "dragStart", "dragEnd"], exportAs: ["matSliderThumb"] }, { type: i4.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { type: i4.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { type: i4.FormControlDirective, selector: "[formControl]", inputs: ["formControl", "disabled", "ngModel"], outputs: ["ngModelChange"], exportAs: ["ngForm"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: FormlyFieldMDCSlider, decorators: [{
            type: Component,
            args: [{
                    selector: 'formly-field-mat-slider',
                    template: `
    <mat-slider
      [id]="id"
      [style.width]="'100%'"
      [formlyAttributes]="field"
      [tabIndex]="props.tabindex"
      [color]="props.color"
      [displayWith]="props.displayWith"
      [invert]="props.invert"
      [max]="props.max"
      [min]="props.min"
      [step]="props.step"
      [discrete]="props.thumbLabel || props.discrete"
      [showTickMarks]="props.showTickMarks"
      [step]="props.step"
      [tickInterval]="props.tickInterval"
      [valueText]="props.valueText"
      [vertical]="props.vertical"
      (input)="props.input && props.input(field, $event)"
      (change)="props.change && props.change(field, $event)"
    >
      <input matSliderThumb [formControl]="formControl" [formlyAttributes]="field" />
    </mat-slider>
  `,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                }]
        }], propDecorators: { sliderThumb: [{
                type: ViewChild,
                args: [MAT_SLIDER_THUMB, { static: true }]
            }], mdcSlider: [{
                type: ViewChild,
                args: [MatSlider, { static: true }]
            }] } });
