// ng-primitives/esm2022/tooltip/config/tooltip.config.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { InjectionToken, inject } from '@angular/core';
export const defaultTooltipConfig = {
    offset: 4,
    placement: 'top',
    showDelay: 0,
    hideDelay: 0,
    flip: true,
    container: null,
};
export const NgpTooltipConfigToken = new InjectionToken('NgpTooltipConfigToken');
/**
 * Provide the default Tooltip configuration
 * @param config The Tooltip configuration
 * @returns The provider
 */
export function provideTooltipConfig(config) {
    return [
        {
            provide: NgpTooltipConfigToken,
            useValue: { ...defaultTooltipConfig, ...config },
        },
    ];
}
/**
 * Inject the Tooltip configuration
 * @returns The global Tooltip configuration
 */
export function injectTooltipConfig() {
    return inject(NgpTooltipConfigToken, { optional: true }) ?? defaultTooltipConfig;
}
