// @ng-matero/extensions/esm2022/datetimepicker/datetimepicker-module.mjs
import { A11yModule } from '@angular/cdk/a11y';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatButtonModule } from '@angular/material/button';
import { MtxCalendar } from './calendar';
import { MtxCalendarBody } from './calendar-body';
import { MtxClock } from './clock';
import { MTX_DATETIMEPICKER_SCROLL_STRATEGY_FACTORY_PROVIDER, MtxDatetimepicker, MtxDatetimepickerContent, } from './datetimepicker';
import { MtxDatetimepickerInput } from './datetimepicker-input';
import { MtxDatetimepickerToggle, MtxDatetimepickerToggleIcon } from './datetimepicker-toggle';
import { MtxMonthView } from './month-view';
import { MtxMultiYearView } from './multi-year-view';
import { MtxTime, MtxTimeInput } from './time';
import { MtxYearView } from './year-view';
import { MtxDatetimepickerActions, MtxDatetimepickerApply, MtxDatetimepickerCancel, MtxDatetimepickerClear, } from './datetimepicker-actions';
import * as i0 from "@angular/core";
export class MtxDatetimepickerModule {
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxDatetimepickerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.0", ngImport: i0, type: MtxDatetimepickerModule, imports: [CommonModule,
            OverlayModule,
            A11yModule,
            PortalModule,
            MatButtonModule,
            MtxCalendar,
            MtxCalendarBody,
            MtxClock,
            MtxTime,
            MtxTimeInput,
            MtxDatetimepicker,
            MtxDatetimepickerToggle,
            MtxDatetimepickerToggleIcon,
            MtxDatetimepickerInput,
            MtxDatetimepickerContent,
            MtxMonthView,
            MtxYearView,
            MtxMultiYearView,
            MtxDatetimepickerApply,
            MtxDatetimepickerCancel,
            MtxDatetimepickerClear,
            MtxDatetimepickerActions], exports: [MtxCalendar,
            MtxCalendarBody,
            MtxClock,
            MtxTime,
            MtxDatetimepicker,
            MtxDatetimepickerToggle,
            MtxDatetimepickerToggleIcon,
            MtxDatetimepickerInput,
            MtxDatetimepickerContent,
            MtxMonthView,
            MtxYearView,
            MtxMultiYearView,
            MtxDatetimepickerApply,
            MtxDatetimepickerCancel,
            MtxDatetimepickerClear,
            MtxDatetimepickerActions] }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxDatetimepickerModule, providers: [MTX_DATETIMEPICKER_SCROLL_STRATEGY_FACTORY_PROVIDER], imports: [CommonModule,
            OverlayModule,
            A11yModule,
            PortalModule,
            MatButtonModule,
            MtxCalendar,
            MtxTime,
            MtxDatetimepickerToggle,
            MtxDatetimepickerContent] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.0", ngImport: i0, type: MtxDatetimepickerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        OverlayModule,
                        A11yModule,
                        PortalModule,
                        MatButtonModule,
                        MtxCalendar,
                        MtxCalendarBody,
                        MtxClock,
                        MtxTime,
                        MtxTimeInput,
                        MtxDatetimepicker,
                        MtxDatetimepickerToggle,
                        MtxDatetimepickerToggleIcon,
                        MtxDatetimepickerInput,
                        MtxDatetimepickerContent,
                        MtxMonthView,
                        MtxYearView,
                        MtxMultiYearView,
                        MtxDatetimepickerApply,
                        MtxDatetimepickerCancel,
                        MtxDatetimepickerClear,
                        MtxDatetimepickerActions,
                    ],
                    exports: [
                        MtxCalendar,
                        MtxCalendarBody,
                        MtxClock,
                        MtxTime,
                        MtxDatetimepicker,
                        MtxDatetimepickerToggle,
                        MtxDatetimepickerToggleIcon,
                        MtxDatetimepickerInput,
                        MtxDatetimepickerContent,
                        MtxMonthView,
                        MtxYearView,
                        MtxMultiYearView,
                        MtxDatetimepickerApply,
                        MtxDatetimepickerCancel,
                        MtxDatetimepickerClear,
                        MtxDatetimepickerActions,
                    ],
                    providers: [MTX_DATETIMEPICKER_SCROLL_STRATEGY_FACTORY_PROVIDER],
                }]
        }] });
