// ng-primitives/esm2022/date-picker/date-picker-previous-month/date-picker-previous-month.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { computed, Directive, ElementRef, HostListener, inject } from '@angular/core';
import { NgpButton } from 'ng-primitives/button';
import { injectDateAdapter } from 'ng-primitives/date-time';
import { NgpDisabledToken } from 'ng-primitives/internal';
import { injectDatePicker } from '../date-picker/date-picker.token';
import { NgpDatePickerPreviousMonthToken } from './date-picker-previous-month.token';
import * as i0 from "@angular/core";
import * as i1 from "ng-primitives/button";
export class NgpDatePickerPreviousMonth {
    constructor() {
        /**
         * Access the element ref.
         */
        this.elementRef = inject(ElementRef);
        /**
         * Access the date adapter.
         */
        this.dateAdapter = injectDateAdapter();
        /**
         * Access the date picker.
         */
        this.datePicker = injectDatePicker();
        /**
         * Determine if this is a button element
         */
        this.isButton = this.elementRef.nativeElement.tagName.toLowerCase() === 'button';
        /**
         * Determine if the next month is disabled.
         * @internal
         */
        this.disabled = computed(() => {
            if (this.datePicker.disabled()) {
                return true;
            }
            const minDate = this.datePicker.min();
            // if the next month is out of bounds, disable it.
            const firstDay = this.dateAdapter.set(this.dateAdapter.startOfMonth(this.datePicker.focusedDate()), {
                hour: 0,
                minute: 0,
                second: 0,
                millisecond: 0,
            });
            // if there is a min date and it is equal to or after the first day of the month, disable it.
            if (minDate && this.dateAdapter.compare(minDate, firstDay) >= 0) {
                return true;
            }
            return false;
        });
    }
    /**
     * Navigate to the previous month.
     */
    navigateToPreviouMonth() {
        if (this.disabled()) {
            return;
        }
        // move focus to the first day of the previous month.
        let date = this.datePicker.focusedDate();
        date = this.dateAdapter.subtract(date, { months: 1 });
        date = this.dateAdapter.set(date, {
            day: 1,
            hour: 0,
            minute: 0,
            second: 0,
            millisecond: 0,
        });
        this.datePicker.setFocusedDate(date, 'mouse', 'backward');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDatePickerPreviousMonth, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.13", type: NgpDatePickerPreviousMonth, isStandalone: true, selector: "[ngpDatePickerPreviousMonth]", host: { listeners: { "click": "navigateToPreviouMonth()" }, properties: { "attr.data-disabled": "disabled() ? \"\" : null", "attr.aria-disabled": "disabled()", "attr.disabled": "isButton && disabled() ? true : null", "attr.type": "isButton ? \"button\" : null" } }, providers: [
            { provide: NgpDatePickerPreviousMonthToken, useExisting: NgpDatePickerPreviousMonth },
            { provide: NgpDisabledToken, useExisting: NgpDatePickerPreviousMonth },
        ], exportAs: ["ngpDatePickerPreviousMonth"], hostDirectives: [{ directive: i1.NgpButton }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpDatePickerPreviousMonth, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpDatePickerPreviousMonth]',
                    exportAs: 'ngpDatePickerPreviousMonth',
                    providers: [
                        { provide: NgpDatePickerPreviousMonthToken, useExisting: NgpDatePickerPreviousMonth },
                        { provide: NgpDisabledToken, useExisting: NgpDatePickerPreviousMonth },
                    ],
                    hostDirectives: [NgpButton],
                    host: {
                        '[attr.data-disabled]': 'disabled() ? "" : null',
                        '[attr.aria-disabled]': 'disabled()',
                        '[attr.disabled]': 'isButton && disabled() ? true : null',
                        '[attr.type]': 'isButton ? "button" : null',
                    },
                }]
        }], propDecorators: { navigateToPreviouMonth: [{
                type: HostListener,
                args: ['click']
            }] } });
