// ng-primitives/esm2022/accordion/accordion-content/accordion-content.directive.mjs
/**
 * Copyright © 2024 Angular Primitives.
 * https://github.com/ng-primitives/ng-primitives
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { Directive, input } from '@angular/core';
import { uniqueId } from 'ng-primitives/utils';
import { injectAccordionItem } from '../accordion-item/accordion-item.token';
import { injectAccordion } from '../accordion/accordion.token';
import { NgpAccordionContentToken } from './accordion-content.token';
import * as i0 from "@angular/core";
export class NgpAccordionContent {
    constructor() {
        /**
         * Access the accordion
         */
        this.accordion = injectAccordion();
        /**
         * Access the accordion item
         */
        this.accordionItem = injectAccordionItem();
        /**
         * The id of the content region
         */
        this.id = input(uniqueId('ngp-accordion-content'));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAccordionContent, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "18.2.13", type: NgpAccordionContent, isStandalone: true, selector: "[ngpAccordionContent]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "region" }, properties: { "id": "id()", "attr.data-orientation": "accordion.orientation()", "attr.data-open": "accordionItem.open() ? \"\" : null", "attr.aria-labelledby": "accordionItem.triggerId()" } }, providers: [{ provide: NgpAccordionContentToken, useExisting: NgpAccordionContent }], exportAs: ["ngpAccordionContent"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: NgpAccordionContent, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngpAccordionContent]',
                    exportAs: 'ngpAccordionContent',
                    providers: [{ provide: NgpAccordionContentToken, useExisting: NgpAccordionContent }],
                    host: {
                        role: 'region',
                        '[id]': 'id()',
                        '[attr.data-orientation]': 'accordion.orientation()',
                        '[attr.data-open]': 'accordionItem.open() ? "" : null',
                        '[attr.aria-labelledby]': 'accordionItem.triggerId()',
                    },
                }]
        }] });
